<!DOCTYPE html>
<!--[if IE 7]><html class="layout-wrapper ie ie7 lte9 lte8 lte7"><![endif]-->
<!--[if IE 8]><html class="layout-wrapper ie ie8 lte9 lte8"><![endif]-->
<!--[if IE 9]><html class="layout-wrapper ie ie9 lte9"><![endif]-->
<!--[if (gte IE 9)|!(IE)]<!-->
<html class="layout-wrapper js flexbox flexboxlegacy canvas canvastext webgl no-touch geolocation postmessage no-websqldatabase indexeddb hashchange history draganddrop websockets rgba hsla multiplebgs backgroundsize borderimage borderradius boxshadow textshadow opacity cssanimations csscolumns cssgradients no-cssreflections csstransforms csstransforms3d csstransitions fontface generatedcontent video audio localstorage sessionstorage webworkers applicationcache svg inlinesvg smil svgclippaths wf-active" style=""><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8"><title>Explore the Human Genome with the SciPy Stack | Toptal</title>
<meta name="description" content="Genome data is one of the most widely analyzed datasets in the realm of Bioinformatics. The SciPy stack offers a suite of popular Python packages designed for numerical computing, data transformation, analysis and visualization, which is ideal for many bioinformatic analysis needs. In this tutorial, Toptal Software Engineer Zhuyi Xue walks us through some of the capabilities of the SciPy stack. He also answers some interesting questions about the human genome, including: How much of the genome is incomplete? How long is a typical gene?">
<link rel="alternate" href="https://www.toptal.com/python/comprehensive-introduction-your-genome-scipy" hreflang="en">
<link rel="canonical" href="https://www.toptal.com/python/comprehensive-introduction-your-genome-scipy">
<meta property="og:image" content="https://uploads.toptal.io/blog/post_image/120208/default-social-5-f95901aa20eb4661d6c68b58abcff464.png">
<meta property="og:title" content="A Comprehensive Introduction To Your Genome With the SciPy Stack">
<meta property="og:description" content="Genome data is one of the most widely analyzed datasets in the realm of Bioinformatics. The SciPy stack offers a suite of popular Python packages designed for numerical computing, data transformation, analysis and visualization, which is ideal for many bioinformatic analysis needs.

In this tutorial, Toptal Software Engineer Zhuyi Xue walks us through some of the capabilities of the SciPy stack. He also answers some interesting questions about the human genome, including: How much of the genome is incomplete? How long is a typical gene?">
<meta property="og:site_name" content="Toptal Engineering Blog">
<meta name="twitter:card" content="summary_large_image"><meta charset="utf-8"><meta content="IE=edge,chrome=1" http-equiv="X-UA-Compatible">
<script type="text/javascript" async="" src="https://a.quora.com/qevents.js"></script><script type="text/javascript" async="" src="https://www.googleadservices.com/pagead/conversion_async.js"></script><script type="text/javascript" async="" src="analytics_002.js"></script><script type="text/javascript" async="" src="bat.js"></script><script type="text/javascript" async="" src="insight.js"></script><script type="text/javascript" async="" src="5977.js"></script><script src="2799924_002.js" type="text/javascript" id="hs-analytics"></script><script type="text/javascript" async="" src="insight.js"></script><script async="" src="analytics.txt"></script><script src="https://connect.facebook.net/signals/config/463369723801939?v=2.8.47&amp;r=stable" async=""></script><script async="" src="fbevents.js"></script><script type="text/javascript" async="" src="analytics.js"></script><script async="" src="widgets.js"></script><script async="" src="gtm.js"></script><script>window.NREUM||(NREUM={});NREUM.info={"beacon":"bam.nr-data.net","errorBeacon":"bam.nr-data.net","licenseKey":"e3359cee6b","applicationID":"30100511","transactionName":"JV0NEUYJCQkARxgVEwReCgYbFgoWEUYYFg4JRQ==","queueTime":0,"applicationTime":47,"agent":""}</script>
<script>(window.NREUM||(NREUM={})).loader_config={xpid:"VgcEVlJaGwAAVVFSAgAG"};window.NREUM||(NREUM={}),__nr_require=function(t,n,e){function r(e){if(!n[e]){var o=n[e]={exports:{}};t[e][0].call(o.exports,function(n){var o=t[e][1][n];return r(o||n)},o,o.exports)}return n[e].exports}if("function"==typeof __nr_require)return __nr_require;for(var o=0;o<e.length;o++)r(e[o]);return r}({1:[function(t,n,e){function r(t){try{s.console&&console.log(t)}catch(n){}}var o,i=t("ee"),a=t(16),s={};try{o=localStorage.getItem("__nr_flags").split(","),console&&"function"==typeof console.log&&(s.console=!0,o.indexOf("dev")!==-1&&(s.dev=!0),o.indexOf("nr_dev")!==-1&&(s.nrDev=!0))}catch(c){}s.nrDev&&i.on("internal-error",function(t){r(t.stack)}),s.dev&&i.on("fn-err",function(t,n,e){r(e.stack)}),s.dev&&(r("NR AGENT IN DEVELOPMENT MODE"),r("flags: "+a(s,function(t,n){return t}).join(", ")))},{}],2:[function(t,n,e){function r(t,n,e,r,s){try{p?p-=1:o(s||new UncaughtException(t,n,e),!0)}catch(f){try{i("ierr",[f,c.now(),!0])}catch(d){}}return"function"==typeof u&&u.apply(this,a(arguments))}function UncaughtException(t,n,e){this.message=t||"Uncaught error with no additional information",this.sourceURL=n,this.line=e}function o(t,n){var e=n?null:c.now();i("err",[t,e])}var i=t("handle"),a=t(17),s=t("ee"),c=t("loader"),f=t("gos"),u=window.onerror,d=!1,l="nr@seenError",p=0;c.features.err=!0,t(1),window.onerror=r;try{throw new Error}catch(h){"stack"in h&&(t(8),t(7),"addEventListener"in window&&t(5),c.xhrWrappable&&t(9),d=!0)}s.on("fn-start",function(t,n,e){d&&(p+=1)}),s.on("fn-err",function(t,n,e){d&&!e[l]&&(f(e,l,function(){return!0}),this.thrown=!0,o(e))}),s.on("fn-end",function(){d&&!this.thrown&&p>0&&(p-=1)}),s.on("internal-error",function(t){i("ierr",[t,c.now(),!0])})},{}],3:[function(t,n,e){t("loader").features.ins=!0},{}],4:[function(t,n,e){function r(t){}if(window.performance&&window.performance.timing&&window.performance.getEntriesByType){var o=t("ee"),i=t("handle"),a=t(8),s=t(7),c="learResourceTimings",f="addEventListener",u="resourcetimingbufferfull",d="bstResource",l="resource",p="-start",h="-end",m="fn"+p,v="fn"+h,w="bstTimer",y="pushState",g=t("loader");g.features.stn=!0,t(6);var b=NREUM.o.EV;o.on(m,function(t,n){var e=t[0];e instanceof b&&(this.bstStart=g.now())}),o.on(v,function(t,n){var e=t[0];e instanceof b&&i("bst",[e,n,this.bstStart,g.now()])}),a.on(m,function(t,n,e){this.bstStart=g.now(),this.bstType=e}),a.on(v,function(t,n){i(w,[n,this.bstStart,g.now(),this.bstType])}),s.on(m,function(){this.bstStart=g.now()}),s.on(v,function(t,n){i(w,[n,this.bstStart,g.now(),"requestAnimationFrame"])}),o.on(y+p,function(t){this.time=g.now(),this.startPath=location.pathname+location.hash}),o.on(y+h,function(t){i("bstHist",[location.pathname+location.hash,this.startPath,this.time])}),f in window.performance&&(window.performance["c"+c]?window.performance[f](u,function(t){i(d,[window.performance.getEntriesByType(l)]),window.performance["c"+c]()},!1):window.performance[f]("webkit"+u,function(t){i(d,[window.performance.getEntriesByType(l)]),window.performance["webkitC"+c]()},!1)),document[f]("scroll",r,{passive:!0}),document[f]("keypress",r,!1),document[f]("click",r,!1)}},{}],5:[function(t,n,e){function r(t){for(var n=t;n&&!n.hasOwnProperty(u);)n=Object.getPrototypeOf(n);n&&o(n)}function o(t){s.inPlace(t,[u,d],"-",i)}function i(t,n){return t[1]}var a=t("ee").get("events"),s=t(19)(a,!0),c=t("gos"),f=XMLHttpRequest,u="addEventListener",d="removeEventListener";n.exports=a,"getPrototypeOf"in Object?(r(document),r(window),r(f.prototype)):f.prototype.hasOwnProperty(u)&&(o(window),o(f.prototype)),a.on(u+"-start",function(t,n){var e=t[1],r=c(e,"nr@wrapped",function(){function t(){if("function"==typeof e.handleEvent)return e.handleEvent.apply(e,arguments)}var n={object:t,"function":e}[typeof e];return n?s(n,"fn-",null,n.name||"anonymous"):e});this.wrapped=t[1]=r}),a.on(d+"-start",function(t){t[1]=this.wrapped||t[1]})},{}],6:[function(t,n,e){var r=t("ee").get("history"),o=t(19)(r);n.exports=r,o.inPlace(window.history,["pushState","replaceState"],"-")},{}],7:[function(t,n,e){var r=t("ee").get("raf"),o=t(19)(r),i="equestAnimationFrame";n.exports=r,o.inPlace(window,["r"+i,"mozR"+i,"webkitR"+i,"msR"+i],"raf-"),r.on("raf-start",function(t){t[0]=o(t[0],"fn-")})},{}],8:[function(t,n,e){function r(t,n,e){t[0]=a(t[0],"fn-",null,e)}function o(t,n,e){this.method=e,this.timerDuration=isNaN(t[1])?0:+t[1],t[0]=a(t[0],"fn-",this,e)}var i=t("ee").get("timer"),a=t(19)(i),s="setTimeout",c="setInterval",f="clearTimeout",u="-start",d="-";n.exports=i,a.inPlace(window,[s,"setImmediate"],s+d),a.inPlace(window,[c],c+d),a.inPlace(window,[f,"clearImmediate"],f+d),i.on(c+u,r),i.on(s+u,o)},{}],9:[function(t,n,e){function r(t,n){d.inPlace(n,["onreadystatechange"],"fn-",s)}function o(){var t=this,n=u.context(t);t.readyState>3&&!n.resolved&&(n.resolved=!0,u.emit("xhr-resolved",[],t)),d.inPlace(t,y,"fn-",s)}function i(t){g.push(t),h&&(x?x.then(a):v?v(a):(E=-E,O.data=E))}function a(){for(var t=0;t<g.length;t++)r([],g[t]);g.length&&(g=[])}function s(t,n){return n}function c(t,n){for(var e in t)n[e]=t[e];return n}t(5);var f=t("ee"),u=f.get("xhr"),d=t(19)(u),l=NREUM.o,p=l.XHR,h=l.MO,m=l.PR,v=l.SI,w="readystatechange",y=["onload","onerror","onabort","onloadstart","onloadend","onprogress","ontimeout"],g=[];n.exports=u;var b=window.XMLHttpRequest=function(t){var n=new p(t);try{u.emit("new-xhr",[n],n),n.addEventListener(w,o,!1)}catch(e){try{u.emit("internal-error",[e])}catch(r){}}return n};if(c(p,b),b.prototype=p.prototype,d.inPlace(b.prototype,["open","send"],"-xhr-",s),u.on("send-xhr-start",function(t,n){r(t,n),i(n)}),u.on("open-xhr-start",r),h){var x=m&&m.resolve();if(!v&&!m){var E=1,O=document.createTextNode(E);new h(a).observe(O,{characterData:!0})}}else f.on("fn-end",function(t){t[0]&&t[0].type===w||a()})},{}],10:[function(t,n,e){function r(t){var n=this.params,e=this.metrics;if(!this.ended){this.ended=!0;for(var r=0;r<d;r++)t.removeEventListener(u[r],this.listener,!1);if(!n.aborted){if(e.duration=a.now()-this.startTime,4===t.readyState){n.status=t.status;var i=o(t,this.lastSize);if(i&&(e.rxSize=i),this.sameOrigin){var c=t.getResponseHeader("X-NewRelic-App-Data");c&&(n.cat=c.split(", ").pop())}}else n.status=0;e.cbTime=this.cbTime,f.emit("xhr-done",[t],t),s("xhr",[n,e,this.startTime])}}}function o(t,n){var e=t.responseType;if("json"===e&&null!==n)return n;var r="arraybuffer"===e||"blob"===e||"json"===e?t.response:t.responseText;return h(r)}function i(t,n){var e=c(n),r=t.params;r.host=e.hostname+":"+e.port,r.pathname=e.pathname,t.sameOrigin=e.sameOrigin}var a=t("loader");if(a.xhrWrappable){var s=t("handle"),c=t(11),f=t("ee"),u=["load","error","abort","timeout"],d=u.length,l=t("id"),p=t(14),h=t(13),m=window.XMLHttpRequest;a.features.xhr=!0,t(9),f.on("new-xhr",function(t){var n=this;n.totalCbs=0,n.called=0,n.cbTime=0,n.end=r,n.ended=!1,n.xhrGuids={},n.lastSize=null,p&&(p>34||p<10)||window.opera||t.addEventListener("progress",function(t){n.lastSize=t.loaded},!1)}),f.on("open-xhr-start",function(t){this.params={method:t[0]},i(this,t[1]),this.metrics={}}),f.on("open-xhr-end",function(t,n){"loader_config"in NREUM&&"xpid"in NREUM.loader_config&&this.sameOrigin&&n.setRequestHeader("X-NewRelic-ID",NREUM.loader_config.xpid)}),f.on("send-xhr-start",function(t,n){var e=this.metrics,r=t[0],o=this;if(e&&r){var i=h(r);i&&(e.txSize=i)}this.startTime=a.now(),this.listener=function(t){try{"abort"===t.type&&(o.params.aborted=!0),("load"!==t.type||o.called===o.totalCbs&&(o.onloadCalled||"function"!=typeof n.onload))&&o.end(n)}catch(e){try{f.emit("internal-error",[e])}catch(r){}}};for(var s=0;s<d;s++)n.addEventListener(u[s],this.listener,!1)}),f.on("xhr-cb-time",function(t,n,e){this.cbTime+=t,n?this.onloadCalled=!0:this.called+=1,this.called!==this.totalCbs||!this.onloadCalled&&"function"==typeof e.onload||this.end(e)}),f.on("xhr-load-added",function(t,n){var e=""+l(t)+!!n;this.xhrGuids&&!this.xhrGuids[e]&&(this.xhrGuids[e]=!0,this.totalCbs+=1)}),f.on("xhr-load-removed",function(t,n){var e=""+l(t)+!!n;this.xhrGuids&&this.xhrGuids[e]&&(delete this.xhrGuids[e],this.totalCbs-=1)}),f.on("addEventListener-end",function(t,n){n instanceof m&&"load"===t[0]&&f.emit("xhr-load-added",[t[1],t[2]],n)}),f.on("removeEventListener-end",function(t,n){n instanceof m&&"load"===t[0]&&f.emit("xhr-load-removed",[t[1],t[2]],n)}),f.on("fn-start",function(t,n,e){n instanceof m&&("onload"===e&&(this.onload=!0),("load"===(t[0]&&t[0].type)||this.onload)&&(this.xhrCbStart=a.now()))}),f.on("fn-end",function(t,n){this.xhrCbStart&&f.emit("xhr-cb-time",[a.now()-this.xhrCbStart,this.onload,n],n)})}},{}],11:[function(t,n,e){n.exports=function(t){var n=document.createElement("a"),e=window.location,r={};n.href=t,r.port=n.port;var o=n.href.split("://");!r.port&&o[1]&&(r.port=o[1].split("/")[0].split("@").pop().split(":")[1]),r.port&&"0"!==r.port||(r.port="https"===o[0]?"443":"80"),r.hostname=n.hostname||e.hostname,r.pathname=n.pathname,r.protocol=o[0],"/"!==r.pathname.charAt(0)&&(r.pathname="/"+r.pathname);var i=!n.protocol||":"===n.protocol||n.protocol===e.protocol,a=n.hostname===document.domain&&n.port===e.port;return r.sameOrigin=i&&(!n.hostname||a),r}},{}],12:[function(t,n,e){function r(){}function o(t,n,e){return function(){return i(t,[f.now()].concat(s(arguments)),n?null:this,e),n?void 0:this}}var i=t("handle"),a=t(16),s=t(17),c=t("ee").get("tracer"),f=t("loader"),u=NREUM;"undefined"==typeof window.newrelic&&(newrelic=u);var d=["setPageViewName","setCustomAttribute","setErrorHandler","finished","addToTrace","inlineHit","addRelease"],l="api-",p=l+"ixn-";a(d,function(t,n){u[n]=o(l+n,!0,"api")}),u.addPageAction=o(l+"addPageAction",!0),u.setCurrentRouteName=o(l+"routeName",!0),n.exports=newrelic,u.interaction=function(){return(new r).get()};var h=r.prototype={createTracer:function(t,n){var e={},r=this,o="function"==typeof n;return i(p+"tracer",[f.now(),t,e],r),function(){if(c.emit((o?"":"no-")+"fn-start",[f.now(),r,o],e),o)try{return n.apply(this,arguments)}catch(t){throw c.emit("fn-err",[arguments,this,t],e),t}finally{c.emit("fn-end",[f.now()],e)}}}};a("actionText,setName,setAttribute,save,ignore,onEnd,getContext,end,get".split(","),function(t,n){h[n]=o(p+n)}),newrelic.noticeError=function(t,n){"string"==typeof t&&(t=new Error(t)),i("err",[t,f.now(),!1,n])}},{}],13:[function(t,n,e){n.exports=function(t){if("string"==typeof t&&t.length)return t.length;if("object"==typeof t){if("undefined"!=typeof ArrayBuffer&&t instanceof ArrayBuffer&&t.byteLength)return t.byteLength;if("undefined"!=typeof Blob&&t instanceof Blob&&t.size)return t.size;if(!("undefined"!=typeof FormData&&t instanceof FormData))try{return JSON.stringify(t).length}catch(n){return}}}},{}],14:[function(t,n,e){var r=0,o=navigator.userAgent.match(/Firefox[\/\s](\d+\.\d+)/);o&&(r=+o[1]),n.exports=r},{}],15:[function(t,n,e){function r(t,n){if(!o)return!1;if(t!==o)return!1;if(!n)return!0;if(!i)return!1;for(var e=i.split("."),r=n.split("."),a=0;a<r.length;a++)if(r[a]!==e[a])return!1;return!0}var o=null,i=null,a=/Version\/(\S+)\s+Safari/;if(navigator.userAgent){var s=navigator.userAgent,c=s.match(a);c&&s.indexOf("Chrome")===-1&&s.indexOf("Chromium")===-1&&(o="Safari",i=c[1])}n.exports={agent:o,version:i,match:r}},{}],16:[function(t,n,e){function r(t,n){var e=[],r="",i=0;for(r in t)o.call(t,r)&&(e[i]=n(r,t[r]),i+=1);return e}var o=Object.prototype.hasOwnProperty;n.exports=r},{}],17:[function(t,n,e){function r(t,n,e){n||(n=0),"undefined"==typeof e&&(e=t?t.length:0);for(var r=-1,o=e-n||0,i=Array(o<0?0:o);++r<o;)i[r]=t[n+r];return i}n.exports=r},{}],18:[function(t,n,e){n.exports={exists:"undefined"!=typeof window.performance&&window.performance.timing&&"undefined"!=typeof window.performance.timing.navigationStart}},{}],19:[function(t,n,e){function r(t){return!(t&&t instanceof Function&&t.apply&&!t[a])}var o=t("ee"),i=t(17),a="nr@original",s=Object.prototype.hasOwnProperty,c=!1;n.exports=function(t,n){function e(t,n,e,o){function nrWrapper(){var r,a,s,c;try{a=this,r=i(arguments),s="function"==typeof e?e(r,a):e||{}}catch(f){l([f,"",[r,a,o],s])}u(n+"start",[r,a,o],s);try{return c=t.apply(a,r)}catch(d){throw u(n+"err",[r,a,d],s),d}finally{u(n+"end",[r,a,c],s)}}return r(t)?t:(n||(n=""),nrWrapper[a]=t,d(t,nrWrapper),nrWrapper)}function f(t,n,o,i){o||(o="");var a,s,c,f="-"===o.charAt(0);for(c=0;c<n.length;c++)s=n[c],a=t[s],r(a)||(t[s]=e(a,f?s+o:o,i,s))}function u(e,r,o){if(!c||n){var i=c;c=!0;try{t.emit(e,r,o,n)}catch(a){l([a,e,r,o])}c=i}}function d(t,n){if(Object.defineProperty&&Object.keys)try{var e=Object.keys(t);return e.forEach(function(e){Object.defineProperty(n,e,{get:function(){return t[e]},set:function(n){return t[e]=n,n}})}),n}catch(r){l([r])}for(var o in t)s.call(t,o)&&(n[o]=t[o]);return n}function l(n){try{t.emit("internal-error",n)}catch(e){}}return t||(t=o),e.inPlace=f,e.flag=a,e}},{}],ee:[function(t,n,e){function r(){}function o(t){function n(t){return t&&t instanceof r?t:t?c(t,s,i):i()}function e(e,r,o,i){if(!l.aborted||i){t&&t(e,r,o);for(var a=n(o),s=m(e),c=s.length,f=0;f<c;f++)s[f].apply(a,r);var d=u[g[e]];return d&&d.push([b,e,r,a]),a}}function p(t,n){y[t]=m(t).concat(n)}function h(t,n){var e=y[t];if(e)for(var r=0;r<e.length;r++)e[r]===n&&e.splice(r,1)}function m(t){return y[t]||[]}function v(t){return d[t]=d[t]||o(e)}function w(t,n){f(t,function(t,e){n=n||"feature",g[e]=n,n in u||(u[n]=[])})}var y={},g={},b={on:p,addEventListener:p,removeEventListener:h,emit:e,get:v,listeners:m,context:n,buffer:w,abort:a,aborted:!1};return b}function i(){return new r}function a(){(u.api||u.feature)&&(l.aborted=!0,u=l.backlog={})}var s="nr@context",c=t("gos"),f=t(16),u={},d={},l=n.exports=o();l.backlog=u},{}],gos:[function(t,n,e){function r(t,n,e){if(o.call(t,n))return t[n];var r=e();if(Object.defineProperty&&Object.keys)try{return Object.defineProperty(t,n,{value:r,writable:!0,enumerable:!1}),r}catch(i){}return t[n]=r,r}var o=Object.prototype.hasOwnProperty;n.exports=r},{}],handle:[function(t,n,e){function r(t,n,e,r){o.buffer([t],r),o.emit(t,n,e)}var o=t("ee").get("handle");n.exports=r,r.ee=o},{}],id:[function(t,n,e){function r(t){var n=typeof t;return!t||"object"!==n&&"function"!==n?-1:t===window?0:a(t,i,function(){return o++})}var o=1,i="nr@id",a=t("gos");n.exports=r},{}],loader:[function(t,n,e){function r(){if(!E++){var t=x.info=NREUM.info,n=p.getElementsByTagName("script")[0];if(setTimeout(u.abort,3e4),!(t&&t.licenseKey&&t.applicationID&&n))return u.abort();f(g,function(n,e){t[n]||(t[n]=e)}),c("mark",["onload",a()+x.offset],null,"api");var e=p.createElement("script");e.src="https://"+t.agent,n.parentNode.insertBefore(e,n)}}function o(){"complete"===p.readyState&&i()}function i(){c("mark",["domContent",a()+x.offset],null,"api")}function a(){return O.exists&&performance.now?Math.round(performance.now()):(s=Math.max((new Date).getTime(),s))-x.offset}var s=(new Date).getTime(),c=t("handle"),f=t(16),u=t("ee"),d=t(15),l=window,p=l.document,h="addEventListener",m="attachEvent",v=l.XMLHttpRequest,w=v&&v.prototype;NREUM.o={ST:setTimeout,SI:l.setImmediate,CT:clearTimeout,XHR:v,REQ:l.Request,EV:l.Event,PR:l.Promise,MO:l.MutationObserver};var y=""+location,g={beacon:"bam.nr-data.net",errorBeacon:"bam.nr-data.net",agent:"js-agent.newrelic.com/nr-1118.min.js"},b=v&&w&&w[h]&&!/CriOS/.test(navigator.userAgent),x=n.exports={offset:s,now:a,origin:y,features:{},xhrWrappable:b,userAgent:d};t(12),p[h]?(p[h]("DOMContentLoaded",i,!1),l[h]("load",r,!1)):(p[m]("onreadystatechange",o),l[m]("onload",r)),c("mark",["firstbyte",s],null,"api");var E=0,O=t(18)},{}]},{},["loader",2,10,4,3]);</script><meta content="width=device-width, initial-scale=1.0" name="viewport"><script>(function(w,d,s,l,i){
  w[l]=w[l]||[];
  w[l].push({
    'gtm.start': new Date().getTime(),
    event:'gtm.js'
  });
  var f=d.getElementsByTagName(s)[0],
      j=d.createElement(s),
      dl=l!='dataLayer'?'&l='+l:'';
  j.async=true;
  j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;
  f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','GTM-K4WD79L');</script><script>
  var _rollbarConfig = {
    accessToken: "cc4cbdcfde904714a4fb77535e3229ae",
    captureUncaught: true,
    async: true,
    captureUnhandledRejections: true,
    payload: {
      environment: "production",
      client: {
        javascript: {
          code_version: "65b599e9a9ece62d8bb8513c92224c490fe27e50",
          source_map_enabled: true,
        }
      },
      server: {
        root: "webpack:///./"
      }
    }
  };
  !function(r){function e(n){if(o[n])return o[n].exports;var t=o[n]={exports:{},id:n,loaded:!1};return r[n].call(t.exports,t,t.exports,e),t.loaded=!0,t.exports}var o={};return e.m=r,e.c=o,e.p="",e(0)}([function(r,e,o){"use strict";var n=o(1),t=o(4);_rollbarConfig=_rollbarConfig||{},_rollbarConfig.rollbarJsUrl=_rollbarConfig.rollbarJsUrl||"https://cdnjs.cloudflare.com/ajax/libs/rollbar.js/2.4.4/rollbar.min.js",_rollbarConfig.async=void 0===_rollbarConfig.async||_rollbarConfig.async;var a=n.setupShim(window,_rollbarConfig),l=t(_rollbarConfig);window.rollbar=n.Rollbar,a.loadFull(window,document,!_rollbarConfig.async,_rollbarConfig,l)},function(r,e,o){"use strict";function n(r){return function(){try{return r.apply(this,arguments)}catch(r){try{console.error("[Rollbar]: Internal error",r)}catch(r){}}}}function t(r,e){this.options=r,this._rollbarOldOnError=null;var o=s++;this.shimId=function(){return o},"undefined"!=typeof window&&window._rollbarShims&&(window._rollbarShims[o]={handler:e,messages:[]})}function a(r,e){if(r){var o=e.globalAlias||"Rollbar";if("object"==typeof r[o])return r[o];r._rollbarShims={},r._rollbarWrappedError=null;var t=new p(e);return n(function(){e.captureUncaught&&(t._rollbarOldOnError=r.onerror,i.captureUncaughtExceptions(r,t,!0),i.wrapGlobals(r,t,!0)),e.captureUnhandledRejections&&i.captureUnhandledRejections(r,t,!0);var n=e.autoInstrument;return e.enabled!==!1&&(void 0===n||n===!0||"object"==typeof n&&n.network)&&r.addEventListener&&(r.addEventListener("load",t.captureLoad.bind(t)),r.addEventListener("DOMContentLoaded",t.captureDomContentLoaded.bind(t))),r[o]=t,t})()}}function l(r){return n(function(){var e=this,o=Array.prototype.slice.call(arguments,0),n={shim:e,method:r,args:o,ts:new Date};window._rollbarShims[this.shimId()].messages.push(n)})}var i=o(2),s=0,d=o(3),c=function(r,e){return new t(r,e)},p=function(r){return new d(c,r)};t.prototype.loadFull=function(r,e,o,t,a){var l=function(){var e;if(void 0===r._rollbarDidLoad){e=new Error("rollbar.js did not load");for(var o,n,t,l,i=0;o=r._rollbarShims[i++];)for(o=o.messages||[];n=o.shift();)for(t=n.args||[],i=0;i<t.length;++i)if(l=t[i],"function"==typeof l){l(e);break}}"function"==typeof a&&a(e)},i=!1,s=e.createElement("script"),d=e.getElementsByTagName("script")[0],c=d.parentNode;s.crossOrigin="",s.src=t.rollbarJsUrl,o||(s.async=!0),s.onload=s.onreadystatechange=n(function(){if(!(i||this.readyState&&"loaded"!==this.readyState&&"complete"!==this.readyState)){s.onload=s.onreadystatechange=null;try{c.removeChild(s)}catch(r){}i=!0,l()}}),c.insertBefore(s,d)},t.prototype.wrap=function(r,e,o){try{var n;if(n="function"==typeof e?e:function(){return e||{}},"function"!=typeof r)return r;if(r._isWrap)return r;if(!r._rollbar_wrapped&&(r._rollbar_wrapped=function(){o&&"function"==typeof o&&o.apply(this,arguments);try{return r.apply(this,arguments)}catch(o){var e=o;throw e&&("string"==typeof e&&(e=new String(e)),e._rollbarContext=n()||{},e._rollbarContext._wrappedSource=r.toString(),window._rollbarWrappedError=e),e}},r._rollbar_wrapped._isWrap=!0,r.hasOwnProperty))for(var t in r)r.hasOwnProperty(t)&&(r._rollbar_wrapped[t]=r[t]);return r._rollbar_wrapped}catch(e){return r}};for(var u="log,debug,info,warn,warning,error,critical,global,configure,handleUncaughtException,handleUnhandledRejection,captureEvent,captureDomContentLoaded,captureLoad".split(","),f=0;f<u.length;++f)t.prototype[u[f]]=l(u[f]);r.exports={setupShim:a,Rollbar:p}},function(r,e){"use strict";function o(r,e,o){if(r){var t;if("function"==typeof e._rollbarOldOnError)t=e._rollbarOldOnError;else if(r.onerror){for(t=r.onerror;t._rollbarOldOnError;)t=t._rollbarOldOnError;e._rollbarOldOnError=t}var a=function(){var o=Array.prototype.slice.call(arguments,0);n(r,e,t,o)};o&&(a._rollbarOldOnError=t),r.onerror=a}}function n(r,e,o,n){r._rollbarWrappedError&&(n[4]||(n[4]=r._rollbarWrappedError),n[5]||(n[5]=r._rollbarWrappedError._rollbarContext),r._rollbarWrappedError=null),e.handleUncaughtException.apply(e,n),o&&o.apply(r,n)}function t(r,e,o){if(r){"function"==typeof r._rollbarURH&&r._rollbarURH.belongsToShim&&r.removeEventListener("unhandledrejection",r._rollbarURH);var n=function(r){var o,n,t;try{o=r.reason}catch(r){o=void 0}try{n=r.promise}catch(r){n="[unhandledrejection] error getting `promise` from event"}try{t=r.detail,!o&&t&&(o=t.reason,n=t.promise)}catch(r){t="[unhandledrejection] error getting `detail` from event"}o||(o="[unhandledrejection] error getting `reason` from event"),e&&e.handleUnhandledRejection&&e.handleUnhandledRejection(o,n)};n.belongsToShim=o,r._rollbarURH=n,r.addEventListener("unhandledrejection",n)}}function a(r,e,o){if(r){var n,t,a="EventTarget,Window,Node,ApplicationCache,AudioTrackList,ChannelMergerNode,CryptoOperation,EventSource,FileReader,HTMLUnknownElement,IDBDatabase,IDBRequest,IDBTransaction,KeyOperation,MediaController,MessagePort,ModalWindow,Notification,SVGElementInstance,Screen,TextTrack,TextTrackCue,TextTrackList,WebSocket,WebSocketWorker,Worker,XMLHttpRequest,XMLHttpRequestEventTarget,XMLHttpRequestUpload".split(",");for(n=0;n<a.length;++n)t=a[n],r[t]&&r[t].prototype&&l(e,r[t].prototype,o)}}function l(r,e,o){if(e.hasOwnProperty&&e.hasOwnProperty("addEventListener")){for(var n=e.addEventListener;n._rollbarOldAdd&&n.belongsToShim;)n=n._rollbarOldAdd;var t=function(e,o,t){n.call(this,e,r.wrap(o),t)};t._rollbarOldAdd=n,t.belongsToShim=o,e.addEventListener=t;for(var a=e.removeEventListener;a._rollbarOldRemove&&a.belongsToShim;)a=a._rollbarOldRemove;var l=function(r,e,o){a.call(this,r,e&&e._rollbar_wrapped||e,o)};l._rollbarOldRemove=a,l.belongsToShim=o,e.removeEventListener=l}}r.exports={captureUncaughtExceptions:o,captureUnhandledRejections:t,wrapGlobals:a}},function(r,e){"use strict";function o(r,e){this.impl=r(e,this),this.options=e,n(o.prototype)}function n(r){for(var e=function(r){return function(){var e=Array.prototype.slice.call(arguments,0);if(this.impl[r])return this.impl[r].apply(this.impl,e)}},o="log,debug,info,warn,warning,error,critical,global,configure,handleUncaughtException,handleUnhandledRejection,_createItem,wrap,loadFull,shimId,captureEvent,captureDomContentLoaded,captureLoad".split(","),n=0;n<o.length;n++)r[o[n]]=e(o[n])}o.prototype._swapAndProcessMessages=function(r,e){this.impl=r(this.options);for(var o,n,t;o=e.shift();)n=o.method,t=o.args,this[n]&&"function"==typeof this[n]&&("captureDomContentLoaded"===n||"captureLoad"===n?this[n].apply(this,[t[0],o.ts]):this[n].apply(this,t));return this},r.exports=o},function(r,e){"use strict";r.exports=function(r){return function(e){if(!e&&!window._rollbarInitialized){r=r||{};for(var o,n,t=r.globalAlias||"Rollbar",a=window.rollbar,l=function(r){return new a(r)},i=0;o=window._rollbarShims[i++];)n||(n=o.handler),o.handler._swapAndProcessMessages(l,o.messages);window[t]=n,window._rollbarInitialized=!0}}}}]);
</script>
<link href="https://assets.toptal.io/assets/front/static/favicons/favicon_ca7136.png" rel="icon"><!--[if IE]><link href="https://assets.toptal.io/assets/front/static/favicons/favicon_68b340.ico" rel="shortcut icon" /><![endif]--><link href="https://assets.toptal.io/assets/front/static/favicons/touch_x57_f72647.png" rel="apple-touch-icon-precomposed" sizes="57x57"><link href="https://assets.toptal.io/assets/front/static/favicons/touch_x72_516451.png" rel="apple-touch-icon-precomposed" sizes="72x72"><link href="https://assets.toptal.io/assets/front/static/favicons/touch_x114_1798a8.png" rel="apple-touch-icon-precomposed" sizes="114x114"><link href="https://assets.toptal.io/assets/front/static/favicons/touch_x57_f72647.png" rel="apple-touch-icon-precomposed"><script src="jquery_and_modernizr-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><script>// TODO: Figure out how to override Modernizr test
(function () {
  var isIE, isIE11, noFlexboxClassName;
  isIE = /MSIE/.test(navigator.userAgent) && !/Opera/.test(navigator.userAgent);
  isIE11 = window.MSInputMethodContext && !/\bEdge\/[0-9.]+$/.test(navigator.userAgent);
  if (window.Modernizr && window.Modernizr.flexbox && (isIE || isIE11)) {
    noFlexboxClassName = document.documentElement.className.replace(/\bflexbox\b/, 'no-flexbox ie-flexbox');
    document.documentElement.className = noFlexboxClassName;
  }
})();</script><script>
//<![CDATA[
window.gon={};gon.flash={};gon.env="production";gon.signedin=false;gon.check_session_url="https:\/\/www.toptal.com\/signed_in";gon.shorten_url="https:\/\/www.toptal.com\/shorten_url";gon.geo_target_url="https:\/\/www.toptal.com\/api\/geo_target";gon.recaptcha_site_key="6LfgR1UUAAAAAJ8-zrjvTk18gKat-4CdBPSrUJoR";gon.recaptcha_url="https:\/\/www.google.com\/recaptcha\/api.js";gon.google_api_browser_key="AIzaSyDED9v55b-c1opDYMEDYBkOgDFGlBMICGY";gon.google_js_maps_key="AIzaSyDspLQumHNqilmD5UHddqqpSjwY1AJpVd0";gon.google_places_api_key="AIzaSyBcP-ExZ1pLLRX2Adlc-faZpjwC3raVQI4";gon.ga_settings={"group":"Blog posts"};gon.current_vertical="developers";gon.url_share_counter=0;gon.chameleon_current_experiments={"bounce_modal_blog":{"google_analytics_dimension":"dimension16","variant_key":"switched_off"},"blog_post_ux_experiments":{"google_analytics_dimension":"dimension9","variant_key":"reading_time_and_progress_bar"}};gon.current_skill_name="Python";gon.current_skill_name_overriden_singular="";gon.current_skill_name_overriden_plural="";
//]]>
</script><link rel="stylesheet" media="screen" href="public_styles-65b599e9a9ece62d8bb8513c92224c490fe27e50.css"><link rel="stylesheet" media="screen" href="portal_styles-65b599e9a9ece62d8bb8513c92224c490fe27e50.css"><link rel="stylesheet" media="screen" href="blog_styles-65b599e9a9ece62d8bb8513c92224c490fe27e50.css"><script>if(!window.appinfo) {
  var thens = []
  var catches = []
  var LocalPromise = {
    then: function(callback, error) {
      thens.push(callback)
      if (error) {
        catches.push(error)
      }
    },
    catch: function (error) {
      catches.push(error)
    }
  }
  var resolve = function() {
    var args = arguments
    thens.forEach(function(callback) {
      callback.apply(null, args)
    })
  }
  var reject = function() {
    var args = arguments
    catches.forEach(function(callback) {
      callback.apply(null, args)
    })
  }

  window.onAppInfoLoaded = LocalPromise

  window.onAppinfoLoad = function() {
    document.cookie = 'appinfo_id_status=doing_request;'

    window.appinfo.grab('https://appinfo.toptal.com', {
      data: gon['application_info_data'] || {},
      callback: function(response, error) {
        if (!error) {
          document.cookie = 'appinfo_id_status=request_completed;'

          resolve({
            applicationInfo: response.data
          })
        } else {
          document.cookie = 'appinfo_id_status=request_failed;'
        }
      }
    })
  }

  document.cookie = 'appinfo_id_status=loading_script;'

  var script = document.createElement('script')
  script.type = 'text/javascript'
  script.onError = function() {
    document.cookie = 'appinfo_id_status=script_load_failed;'
    reject()
  }
  script.src = 'https://appinfo.toptal.com/script.js'
  document.getElementsByTagName('head')[0].appendChild(script)
}</script><script type="text/javascript" src="script.js"></script><script type="text/javascript" async="" src="embed.js"></script><style type="text/css"></style><style type="text/css"></style><script type="text/javascript" charset="utf-8" async="" src="3-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><script type="text/javascript" charset="utf-8" async="" src="0-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><link rel="prefetch" href="https://c.disquscdn.com/next/embed/styles/lounge.694ea7181ea49f1ce306dfc00c532f53.css"><link rel="prefetch" href="https://c.disquscdn.com/next/embed/common.bundle.880980e048a2432334f13013030456ac.js"><link rel="prefetch" href="https://c.disquscdn.com/next/embed/lounge.bundle.98707fed95924bce08af95208637b782.js"><link rel="prefetch" href="https://disqus.com/next/config.js"><style type="text/css">.hljs{display:block;overflow-x:auto;padding:.5em;background:#fdf6e3;color:#657b83}.hljs-comment,.hljs-quote{color:#93a1a1}.hljs-addition,.hljs-keyword,.hljs-selector-tag{color:#859900}.hljs-doctag,.hljs-literal,.hljs-meta .hljs-meta-string,.hljs-number,.hljs-regexp,.hljs-string{color:#2aa198}.hljs-name,.hljs-section,.hljs-selector-class,.hljs-selector-id,.hljs-title{color:#268bd2}.hljs-attr,.hljs-attribute,.hljs-class .hljs-title,.hljs-template-variable,.hljs-type,.hljs-variable{color:#b58900}.hljs-bullet,.hljs-link,.hljs-meta,.hljs-meta .hljs-keyword,.hljs-selector-attr,.hljs-selector-pseudo,.hljs-subst,.hljs-symbol{color:#cb4b16}.hljs-built_in,.hljs-deletion{color:#dc322f}.hljs-formula{background:#eee8d5}.hljs-emphasis{font-style:italic}.hljs-strong{font-weight:700}</style><script charset="utf-8" src="button.js"></script></head><body class="layout" data-view="layout#body" id="top"><div class="layout_layer" data-view="layout#layout"><header class="page_header hide_in_webview" data-role="fixed-header"><div class="page_header_line"></div><div class="page_header" data-view="layout#menu"><div class="blog-simple-header grid-row is-light_gray" data-role="main_menu_wrapper"><div class="grid-row-inner is-wide has-padding page_header_top"><div class="page_header_top-logotype_wrapper"><div class="page_header_top-logotype"><a class="logo is-big is-link" href="https://www.toptal.com/developers"></a></div><div class="page_header_top-main_title"><b>Hiring?</b> Toptal handpicks <b><a href="https://www.toptal.com/python" onclick='window.googleAnalytics.trackEvent(event, "BlogCTAs", "ClickedOnHeaderHeadline", "ClickedOnHeaderHeadline")'>top Python engineers</a></b> to suit your needs.</div></div><a class="page_header_top-expand" data-role="expand_menu" href="#"><span class="page_header_top-expand_icon"></span></a><nav class="page_header_top-nav_wrapper"><ul class="page_header_top-nav"><li class="page_header_top-nav_item"><a class="button is-green_candy is-small page_header_top-nav_link is-button" data-role="companies_apply_button" href="https://www.toptal.com/python" onclick='window.googleAnalytics.trackEvent(event, "BlogCTAs", "ClickedOnHeaderCTA", "ClickedOnHeaderCTA");'>Start hiring</a></li><li class="page_header_top-nav_item"><a class="button is-small is-light page_header_top-nav_link is-button" href="https://www.toptal.com/users/login">Log in</a></li></ul></nav></div></div><nav class="grid-row has-top_border is-light_gray page_header_menu-wrapper" data-role="main_menu_wrapper"><div class="grid-row-inner is-wide has-small_padding page_header_menu"><div class="page_header_menu-items-wrapper"><ul class="page_header_menu-items" data-role="menu"><li class="page_header_menu-item is-home"><a class="page_header_menu-link" href="https://www.toptal.com/developers"></a></li><li class="page_header_menu-item is-top-3-percent"><a class="page_header_menu-link" href="https://www.toptal.com/top-3-percent">Top 3%</a></li><li class="page_header_menu-item is-why"><a class="page_header_menu-link" href="https://www.toptal.com/why">Why</a></li><li class="page_header_menu-item is-clients"><a class="page_header_menu-link" href="https://www.toptal.com/clients">Clients</a></li><li class="page_header_menu-item is-enterprise"><a class="page_header_menu-link" href="https://www.toptal.com/enterprise">Enterprise</a></li><li class="page_header_menu-item is-community"><a class="page_header_menu-link" href="https://www.toptal.com/community">Community</a></li><li class="page_header_menu-item is-blog"><a class="page_header_menu-link" href="https://www.toptal.com/developers/blog">Blog</a></li><li class="page_header_menu-item is-about"><a class="page_header_menu-link" href="https://www.toptal.com/about">About Us</a></li><li class="page_header_menu-item is-join"><a class="button is-green_candy is-small page_hader_menu-button" data-role="companies_apply_button" href="https://www.toptal.com/python" onclick='window.googleAnalytics.trackEvent(event, "BlogCTAs", "ClickedOnHeaderCTA", "ClickedOnHeaderCTA");'>Start hiring</a></li><li class="page_header_menu-item is-join"><a class="button is-green_candy is-small page_hader_menu-button" href="https://www.toptal.com/developers/join">Apply as a Developer</a></li><li class="page_header_menu-item is-login"><a class="button is-light is-small page_hader_menu-button is-login" href="https://www.toptal.com/users/login">Login</a></li><li class="page_header_menu-item is-contacts"><ul class="page_header_menu-contact-items is-replaceable"><li class="page_header_menu-contact-item is-contact">Questions?</li><li class="page_header_menu-contact-item is-contact"><a class="link is-dotted page_header_menu-contact-link" href="https://www.toptal.com/contact">Contact Us</a></li><li class="page_header_menu-contact-item"><a class="page_header_menu-contact-link is-social is-facebook" href="https://www.facebook.com/toptal" target="_blank" rel="noopener noreferrer"><span class="page_header_menu-contact-link-icon"></span></a></li><li class="page_header_menu-contact-item"><a class="page_header_menu-contact-link is-social is-twitter" href="https://twitter.com/toptal" target="_blank" rel="noopener noreferrer"><span class="page_header_menu-contact-link-icon"></span></a></li><li class="page_header_menu-contact-item"><a class="page_header_menu-contact-link is-social is-linkedin" href="https://www.linkedin.com/company/toptal" target="_blank" rel="noopener noreferrer"><span class="page_header_menu-contact-link-icon"></span></a></li></ul></li></ul></div><div class="blog_search" data-view="quick_search#search_field" data-vertical="Developer"><form class="blog_search-form" data-role="form"><input autocomplete="off" class="blog_search-field" data-role="field" name="text" placeholder="Search Topics" type="text"><button class="button is-tiny blog_search-submit"><span class="blog_search-submit_icon"></span></button></form><div class="blog_search-suggestions" data-role="suggestions_wrapper"></div><div class="blog_search_results" data-role="results" style="display: none;"></div></div><a class="button is-green_candy is-tiny for-page_header_menu page_header_menu-hire" data-role="companies_apply_button" href="https://www.toptal.com/hire?interested_in=developers&amp;skill=python">Hire a developer</a></div></nav></div></header><main class="layout-main"><div class="reading-progress" data-view="content#reading_progress" style="width: 0px;"></div><header class="grid-row has-top_border has-bottom_border post_cover is-post"><div class="grid-row-inner is-wide has-big_padding blog_title for-show"></div></header><article class="grid-row is-transparent"><div class="grid-row-inner is-wide"><div class="post_title"><div class="post_title-header"><p class="post-reading_time" data-view="content#reading_time" data-minutes="23">23 min read</p><h1 class="post_title-text">A Comprehensive Introduction To Your Genome With the SciPy Stack</h1></div><a class="post_title-back" href="https://www.toptal.com/developers/blog">View all articles</a></div></div></article><section class="grid-row has-top_border is-transparent has-no_border" data-view="" id="main_content"><div class="grid-row-inner has-medium_padding two_columns_layout is-wide"><div class="two_columns_layout-content"><div class="post is-full"><div class="post_meta-wrapper post_header"><div class="post_header-triangle"></div><div class="post_avatar post_meta-avatar"><img class="post_avatar-image" src="small_45dcc56c2c3475b8c81e30a9ae7812c2.jpg" width="50"></div><div class="post_meta is-full"><div class="post_meta-author"><p>by <strong><a class="link is-blue" target="_blank" href="https://www.toptal.com/resume/zhuyi-xue">Zhuyi Xue</a></strong> - Freelance Software Engineer
@ <a class="link is-blue" target="_blank" href="https://www.toptal.com/">Toptal</a>
</p>
</div><div class="post_meta-tags"><a href="https://www.toptal.com/developers/blog/tags/bioinformatics">#Bioinformatics</a> <a href="https://www.toptal.com/developers/blog/tags/genomics">#Genomics</a> <a href="https://www.toptal.com/developers/blog/tags/python">#Python</a> <a href="https://www.toptal.com/developers/blog/tags/scipy">#SciPy</a></div></div><ul class="social_share is-horizontal is-small is-centered_on_mobile for-blog_post_heading is-loaded" data-hashtags="Bioinformatics,Python,SciPy,Genomics" data-url="https://www.toptal.com/python/comprehensive-introduction-your-genome-scipy" data-view="layout#social_share" data-youtube-channel-url="https://www.youtube.com/channel/UCNqm_euTHZz3o5OnKhUS-oA"><li class="social_share-item is-counter" title="Total number of shares"><span class="social_share-item_num" data-role="counter_num">351</span><span class="social_share-item_text" data-role="counter_text">shares</span></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;twitter&quot;}" data-type="twitter" title="Share on Twitter" rel="noopener noreferrer"><img class="social_share-item_image" src="twitter_83c6d4.png" width="50" height="50"></a></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;facebook&quot;}" data-type="facebook" title="Share on Facebook" rel="noopener noreferrer"><img class="social_share-item_image" src="facebook_dc66c9.png" width="50" height="50"></a></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;linkedin&quot;}" data-type="linkedin" title="Share on LinkedIn" rel="noopener noreferrer"><img class="social_share-item_image" src="linkedin_b923c3.png" width="50" height="50"></a></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;pocket&quot;}" data-type="pocket" title="Save to Pocket" rel="noopener noreferrer"><img class="social_share-item_image" src="pocket_e4260c.png" width="50" height="50"></a></li></ul><div class="post_meta-extra"></div></div><div class="content_body" data-insert-middle-content="true" data-view="content#body"><p><a href="https://en.wikipedia.org/wiki/Bioinformatics" rel="noopener noreferrer" target="_blank">Bioinformatics</a> is an interdisciplinary field that develops methods and software tools for analyzing and understanding biological data.</p>

<p>More simply stated, you can simply think of it as <a href="https://www.toptal.com/data-science">data science</a> for biology.</p>

<p>Among the many types of biological data, genomics data is one of the 
most widely analyzed. Especially with the rapid advancement of 
next-generation DNA sequencing (NGS) technologies, the volume of 
genomics data has been growing exponentially. According to <a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1002195" rel="noopener noreferrer" target="_blank">Stephens, Zachary D et al.</a>, genomics data acquisition is on the exabyte-per-year scale.</p>

<p><img src="toptal-blog-image-1478600437768-72348223698908b523c4b825b9e0.png" alt="The Comprehensive Introduction To Your Genome With SciPy"></p>

<div class="pop_out_box is-full_width is-big" style="text-align: center;">SciPy gathers many Python modules for scientific computing, which is ideal for many bioinformatic needs.</div>

<div class="tweet_this">
<a href="https://twitter.com/share" class="twitter-share-button" data-url="https://www.toptal.com/python/comprehensive-introduction-your-genome-scipy" data-text="SciPy gathers many Python modules for scientific computing, which is ideal for many bioinformatic needs." data-via="toptal" rel="noopener noreferrer" target="_blank" data-twitter-extracted-i1554999561626445806="true">Tweet</a>
</div>

<p>In this post, I demo an example of analyzing a GFF3 file for the human genome with the SciPy Stack. <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md" rel="noopener noreferrer" target="_blank">Generic Feature Format Version 3</a>
 (GFF3) is the current standard text file format for storing genomic 
features. In particular, in this post you will learn how to use the 
SciPy stack to answer the following questions about the human genome:</p>

<ol>
  <li>How much of the genome is incomplete?</li>
  <li>How many genes are there in the genome?</li>
  <li>How long is a typical gene?</li>
  <li>What does gene distribution among chromosomes look like?</li>
</ol>

<p>The latest GFF3 file for the human genome can be downloaded from <a href="ftp://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens" rel="noopener noreferrer" target="_blank">here</a>. The <a href="http://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/README" rel="noopener noreferrer" target="_blank">README</a> file that comes in this directory provides a brief description of this data format, and a more thorough specification is found <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md#generic-feature-format-version-3-gff3" rel="noopener noreferrer" target="_blank">here</a>.</p>

<p>We will use <a href="http://pandas.pydata.org/" rel="noopener noreferrer" target="_blank">Pandas</a>,
 a major component of the SciPy stack providing fast, flexible, and 
expressive data structures, to manipulate and understand the GFF3 file.</p>

<h2 id="setup">Setup</h2>

<p>First things first, let’s setup a virtual environment with the SciPy 
stack installed. This process can be time-consuming if built from the 
source manually, as the stack involves many packages – some of which 
depends on external FORTRAN or C code. Here, I recommend using <a href="http://conda.pydata.org/miniconda.html" rel="noopener noreferrer" target="_blank">Miniconda</a>, which makes the setup very easy.</p>

<pre><code class="language-bash hljs">wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh -b
</code></pre>

<p>The <code>-b</code> flag on the bash line tells it to execute in 
batch mode. After the commands above are used to successfully install 
Miniconda, start a new virtual environment for genomics, and then 
install the SciPy stack.</p>

<pre><code class="language-bash hljs">mkdir -p genomics
<span class="hljs-built_in">cd</span> genomics
conda create -p venv ipython matplotlib pandas
</code></pre>

<p>Note that we have only specified the 3 packages we are going to use in this post.
If you want all the packages listed in the SciPy stack, simply append them to the end of the <code>conda create</code> command.</p>

<p>If you are unsure of the exact name of a package, try <code>conda search</code>. Let’s activate the virtual environment and start IPython.</p>

<pre><code class="language-bash hljs"><span class="hljs-built_in">source</span> activate venv/
ipython
</code></pre>

<p><a href="https://ipython.org/" rel="noopener noreferrer" target="_blank">IPython</a> is a significantly more powerful replacement to the default <a href="https://www.toptal.com/python/why-are-there-so-many-pythons">Python interpreter interface</a>,
 so whatever you used to do in the default python interpreter can also 
be done in IPython. I highly recommend every Python programmer, who 
hasn’t been using IPython yet, to give it a try.</p>

<h2 id="download-the-annotation-file">Download the Annotation File</h2>

<p>With our setup now completed, let’s download the human genome annotation file in GFF3 format.</p>

<p>It is about 37 MB, a very small file compared to the information 
content of a human genome, which is about 3 GB in plain text. That’s 
because the GFF3 file only contains the annotation of the sequences, 
while the sequence data is usually stored in another file format called <a href="https://en.wikipedia.org/wiki/FASTA_format" rel="noopener noreferrer" target="_blank">FASTA</a>. If you are interested, you can download FASTA <a href="ftp://ftp.ensembl.org/pub/release-85/fasta/homo_sapiens/dna/" rel="noopener noreferrer" target="_blank">here</a>, but we won’t use the sequence data in this tutorial.</p>

<pre><code class="language-bash hljs">!wget ftp://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/Homo_sapiens.GRCh38.85.gff3.gz
</code></pre>

<p>The prefixed <code>!</code> tells IPython that this is a shell 
command instead of a Python command. However, IPython can also process 
some frequently used shell commands like <code>ls</code>, <code>pwd</code>, <code>rm</code>, <code>mkdir</code>, <code>rmdir</code> even without a prefixed <code>!</code>.</p>

<p>Taking a look at the head of the GFF file, you will see many metadata/pragmas/directives lines starting with <code>##</code> or <code>#!</code>.</p>

<p>According to the <a href="http://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/README" rel="noopener noreferrer" target="_blank">README</a>,  <code>##</code> means the metadata is stable, while <code>#!</code> means it’s experimental.</p>

<p>Later on you will also see <code>###</code>, which is another directive with yet more subtle meaning based on the <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md#generic-feature-format-version-3-gff3" rel="noopener noreferrer" target="_blank">specification</a>.</p>

<p>Human readable comments are supposed to be after a single <code>#</code>. For simplicity, we will treat all lines starting with <code>#</code> as comments, and simply ignore them during our analysis.</p>

<pre><code>##gff-version   3
##sequence-region   1 1 248956422
##sequence-region   10 1 133797422
##sequence-region   11 1 135086622
##sequence-region   12 1 133275309
...
##sequence-region   MT 1 16569
##sequence-region   X 1 156040895
##sequence-region   Y 2781480 56887902
#!genome-build  GRCh38.p7
#!genome-version GRCh38
#!genome-date 2013-12
#!genome-build-accession NCBI:GCA_000001405.22
#!genebuild-last-updated 2016-06
</code></pre>

<p>The first line indicates that the version of GFF format used in this file is 3.</p>

<p>Following that are summaries of all sequence regions. As we will see 
later, such information can also be found in the body part of the file.</p>

<p>The lines starting with <code>#!</code> shows information about the particular build of the genome, GRCh38.p7, that this annotation file applies to.</p>

<p><a href="http://www.ncbi.nlm.nih.gov/projects/genome/assembly/grc/" rel="noopener noreferrer" target="_blank">Genome Reference Consortium</a>
 (GCR) is an international consortium, which oversees updates and 
improvements to several reference genome assemblies, including those for
 human, mouse, zebrafish, and chicken.</p>

<p>Scanning through this file, here are the first few annotation lines.</p>

<pre><code>1       GRCh38  chromosome      1       248956422       .       .       .       ID=chromosome:1;Alias=CM000663.2,chr1,NC_000001.11
###
1       .       biological_region       10469   11240   1.3e+03 .       .       external_name=oe %3D 0.79;logic_name=cpg
1       .       biological_region       10650   10657   0.999   +       .       logic_name=eponine
1       .       biological_region       10655   10657   0.999   -       .       logic_name=eponine
1       .       biological_region       10678   10687   0.999   +       .       logic_name=eponine
1       .       biological_region       10681   10688   0.999   -       .       logic_name=eponine
...
</code></pre>

<p>The columns are <em>seqid</em>, <em>source</em>, <em>type</em>, <em>start</em>, <em>end</em>, <em>score</em>, <em>strand</em>, <em>phase</em>, <em>attributes</em>. Some of them are very easy to understand. Take the first line as an example:</p>

<pre><code>1       GRCh38  chromosome      1       248956422       .       .       .       ID=chromosome:1;Alias=CM000663.2,chr1,NC_000001.11
</code></pre>

<p>This is the annotation of the first chromosome with a seqid of 1, which starts from the first base to the 24,895,622nd base.</p>

<p>In other words, the first chromosome is about 25 million bases long.</p>

<p>Our analysis won’t need information from the three columns with a value of <code>.</code> (i.e. score, strand, and phase), so we can simply ignore them for now.</p>

<p>The last attributes column says Chromosome 1 also has three alias 
names, namely CM000663.2, chr1, and NC_000001.11. That’s basically what a
 GFF3 file looks like, but we won’t inspect them line by line, so it’s 
time to load the whole file into Pandas.</p>

<p>Pandas is good fit for dealing with GFF3 format because it is a 
tab-delimited file, and Pandas has very good support for reading 
CSV-like files.</p>

<p>Note one exception to the tab-delimited format is when the GFF3 contains <code>##FASTA </code>.</p>

<p>According to the <a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md" rel="noopener noreferrer" target="_blank">specification</a>, <code>##FASTA</code>
 indicates the end of an annotation portion, which will be followed with
 one or more sequences in FASTA (a non tab-delimited) format. But this 
is not the case for the GFF3 file we’re going to analyze.</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">1</span>]: <span class="hljs-keyword">import</span> pandas <span class="hljs-keyword">as</span> pd


In [<span class="hljs-number">2</span>]: pd.__version__
Out[<span class="hljs-number">2</span>]: <span class="hljs-string">'0.18.1'</span>


In [<span class="hljs-number">3</span>]: col_names = [<span class="hljs-string">'seqid'</span>, <span class="hljs-string">'source'</span>, <span class="hljs-string">'type'</span>, <span class="hljs-string">'start'</span>, <span class="hljs-string">'end'</span>, <span class="hljs-string">'score'</span>, <span class="hljs-string">'strand'</span>, <span class="hljs-string">'phase'</span>, <span class="hljs-string">'attributes'</span>]
Out[<span class="hljs-number">3</span>]: df = pd.read_csv(<span class="hljs-string">'Homo_sapiens.GRCh38.85.gff3.gz'</span>, compression=<span class="hljs-string">'gzip'</span>,
                         sep=<span class="hljs-string">'\t'</span>, comment=<span class="hljs-string">'#'</span>, low_memory=<span class="hljs-keyword">False</span>,
                         header=<span class="hljs-keyword">None</span>, names=col_names)
</code></pre>

<p>The last line above loads the entire GFF3 file with <code>pandas.read_csv</code> method.</p>

<p>Since it is not a standard CSV file, we need to customize the call a bit.</p>

<p>First, we inform Pandas of the unavailability of header information in the GFF3 with <code>header=None</code>, and then we specify the exact name for each column with <code>names=col_names</code>.</p>

<p>If the <code>names</code> argument is not specified, Pandas will use incremental numbers as names for each column.</p>

<p><code>sep='\t'</code> tells Pandas the columns are tab-separated instead of comma-separated. The value to <code>sep=</code>
 can actually be a regular expression (regex). This becomes handy if the
 file at hand uses different separators for each column (oh yeah, that 
happens). <code>comment='#'</code> means lines starting with <code>#</code> are considered comments and will be ignored.</p>

<p><code>compression='gzip'</code> tells Pandas that the input file is a gzip-compressed.</p>

<p>In addition, <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html" rel="noopener noreferrer" target="_blank"><code>pandas.read_csv</code></a> has a rich set of parameters that allows different kinds of CSV-like file formats to be read.</p>

<p>The type of the returned value is a <code>DataFrame</code>, which is the most important data structure in Pandas, used for representing 2D data.</p>

<p>Pandas also has a <code>Series</code> and <code>Panel</code> data structure for 1D and 3D data, respectively. Please refer to the <a href="http://pandas.pydata.org/pandas-docs/stable/dsintro.html" rel="noopener noreferrer" target="_blank">documentation</a> for an introduction to Pandas’ data structures.</p>

<p>Let’s take a look at the first few entries with <code>.head</code> method.</p>

<pre><code>In [18]: df.head()
Out[18]:
  seqid  source               type  start        end    score strand phase                                         attributes
0     1  GRCh38         chromosome      1  248956422        .      .     .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...
1     1       .  biological_region  10469      11240  1.3e+03      .     .           external_name=oe %3D 0.79;logic_name=cpg
2     1       .  biological_region  10650      10657    0.999      +     .                                 logic_name=eponine
3     1       .  biological_region  10655      10657    0.999      -     .                                 logic_name=eponine
4     1       .  biological_region  10678      10687    0.999      +     .                                 logic_name=eponine
</code></pre>

<p>The output is nicely formatted in a tabular format with longer strings in the attributes column partially replaced with <code>...</code>.</p>

<p>You can set Pandas to not omit long strings with <code>pd.set_option('display.max_colwidth', -1)</code>. In addition, Pandas has many <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.set_option.html" rel="noopener noreferrer" target="_blank">options</a> that can be customized.</p>

<p>Next, let’s get some basic information about this dataframe with the <code>.info</code> method.</p>

<pre><code>In [20]: df.info()
&lt;class 'pandas.core.frame.DataFrame'&gt;
RangeIndex: 2601849 entries, 0 to 2601848
Data columns (total 9 columns):
seqid         object
source        object
type          object
start         int64
end           int64
score         object
strand        object
phase         object
attributes    object
dtypes: int64(2), object(7)
memory usage: 178.7+ MB
</code></pre>

<p>This shows that the GFF3 has 2,601,848 annotated lines, and each line has nine columns.</p>

<p>For each column, it also shows their data types.</p>

<p>That start and end are of int64 type, integers representing positions in the genome.</p>

<p>The other columns are all of type <code>object</code>, which probably means their values consist of a mixture of integers, floats, and strings.</p>

<p>The size of all the information is about 178.7+ MB stored in memory. 
This turns out to be more compact than the uncompressed file, which will
 be about 402 MB. A quick verification is shown below.</p>

<pre><code class="language-bash hljs">gunzip -c Homo_sapiens.GRCh38.85.gff3.gz &gt; /tmp/tmp.gff3 &amp;&amp; du -s /tmp/tmp.gff3
402M    /tmp/tmp.gff3
</code></pre>

<p>From a high-level view, we have loaded the entire GFF3 file into a 
DataFrame object in Python, and all of our following analysis will be 
based on this single object.</p>

<p>Now, let’s see what the first column seqid is all about.</p>

<pre><code>In [29]: df.seqid.unique()
Out[29]:
array(['1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19',
       '2', '20', '21', '22', '3', '4', '5', '6', '7', '8', '9',
       'GL000008.2', 'GL000009.2', 'GL000194.1', 'GL000195.1',
       ...
       'KI270757.1', 'MT', 'X', 'Y'], dtype=object)
In [30]: df.seqid.unique().shape
Out[30]: (194,)
</code></pre>

<p><code>df.seqid</code> is one way to access column data from a dataframe. Another way is <code>df['seqid']</code>, which is more general syntax, because if the column name is a Python reserved keyword (e.g. <code>class</code>) or contains a <code>.</code> or space character, the first way (<code>df.seqid</code>) won’t work.</p>

<p>The output shows that there are 194 unique seqids, which include 
Chromosomes 1 to 22, X, Y, and mitochondrion (MT) DNA as well as 169 
others seqids.</p>

<p>The seqids starting with KI and GL are DNA sequences – or scaffolds –
 in the genome that have not been successfully assembled into the 
chromosomes.</p>

<p>For those who are unfamiliar with genomics, this is important.</p>

<p>Although the first human genome draft came out more than 15 years 
ago, the current human genome is still incomplete. The difficulty in 
assembling these sequences is largely due to <a href="http://www.nature.com/nature/journal/v517/n7536/pdf/nature13907.pdf" rel="noopener noreferrer" target="_blank">complex repetitive regions in the genome</a>.</p>

<p>Next, let’s take a look at the source column.</p>

<p>The <a href="ftp://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/README" rel="noopener noreferrer" target="_blank">README</a>
 says that the source is a free text qualifier intended to describe the 
algorithm or operating procedure that generated this feature.</p>

<pre><code>In [66]: df.source.value_counts()
Out[66]:
havana            1441093
ensembl_havana     745065
ensembl            228212
.                  182510
mirbase              4701
GRCh38                194
insdc                  74
</code></pre>

<p>This is an example of the use of the <code>value_counts</code> method, which is extremely useful for a quick count of categorical variables.</p>

<p>From the result, we can see that there are seven possible values for 
this column, and the majority of entries in the GFF3 file come from 
havana, ensembl and ensembl_havana.</p>

<p>You can learn more about what these sources mean and the relationships between them in <a href="http://uswest.ensembl.org/Help/Faq?id=152" rel="noopener noreferrer" target="_blank">this post</a>.</p>

<p>To keep things simple, we will focus on entries from sources GRCh38, havana, ensembl, and ensembl_havan.a.</p>

<h4 id="how-much-of-the-genome-is-incomplete">How Much of the Genome Is Incomplete?</h4>

<p>The information about each entire chromosome is in the entries from 
source GRCh38, so let’s first filter out the rest, and assign the 
filtered result to a new variable <code>gdf</code>.</p>

<pre><code>In [70]: gdf = df[df.source == 'GRCh38']
In [87]: gdf.shape
Out[87]: (194, 9)
In [84]: gdf.sample(10)
Out[84]:
              seqid  source         type  start        end score strand phase                                         attributes
2511585  KI270708.1  GRCh38  supercontig      1     127682     .      .     .  ID=supercontig:KI270708.1;Alias=chr1_KI270708v...
2510840  GL000208.1  GRCh38  supercontig      1      92689     .      .     .  ID=supercontig:GL000208.1;Alias=chr5_GL000208v...
990810           17  GRCh38   chromosome      1   83257441     .      .     .  ID=chromosome:17;Alias=CM000679.2,chr17,NC_000...
2511481  KI270373.1  GRCh38  supercontig      1       1451     .      .     .  ID=supercontig:KI270373.1;Alias=chrUn_KI270373...
2511490  KI270384.1  GRCh38  supercontig      1       1658     .      .     .  ID=supercontig:KI270384.1;Alias=chrUn_KI270384...
2080148           6  GRCh38   chromosome      1  170805979     .      .     .  ID=chromosome:6;Alias=CM000668.2,chr6,NC_00000...
2511504  KI270412.1  GRCh38  supercontig      1       1179     .      .     .  ID=supercontig:KI270412.1;Alias=chrUn_KI270412...
1201561          19  GRCh38   chromosome      1   58617616     .      .     .  ID=chromosome:19;Alias=CM000681.2,chr19,NC_000...
2511474  KI270340.1  GRCh38  supercontig      1       1428     .      .     .  ID=supercontig:KI270340.1;Alias=chrUn_KI270340...
2594560           Y  GRCh38   chromosome  2781480  56887902     .      .     .  ID=chromosome:Y;Alias=CM000686.2,chrY,NC_00002...
</code></pre>

<p>Filtering is easy in Pandas.</p>

<p>If you inspect the value evaluated from the expression <code>df.source == 'GRCh38'</code>, it’s a series of <code>True</code> and <code>False</code> values for each entry with the same index as <code>df</code>. Passing it to <code>df[]</code> will only return those entries where their corresponding values are True.</p>

<p>There are 194 keys in <code>df[]</code> for which <code>df.source == 'GRCh38'</code>.</p>

<p>As we’ve seen previously, there are also 194 unique values in the <code>seqid</code> column, meaning each entry in <code>gdf</code> corresponds to a particular seqid.</p>

<p>Then we randomly select 10 entries with the <code>sample</code> method to take closer look.</p>

<p>You can see that the unassembled sequences are of type supercontig 
while the others are of chromosome. To compute the fraction of genome 
that’s incomplete, we first need to know the length of the entire 
genome, which is the sum of the lengths of all seqids.</p>

<pre><code>In [90]: gdf = gdf.copy()
In [91]: gdf['length'] = gdf.end - gdf.start + 1
In [93]: gdf.head()
Out[93]:
       seqid  source        type  start        end score strand phase                                         attributes     length
0          1  GRCh38  chromosome      1  248956422     .      .     .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  248956421
235068    10  GRCh38  chromosome      1  133797422     .      .     .  ID=chromosome:10;Alias=CM000672.2,chr10,NC_000...  133797421
328938    11  GRCh38  chromosome      1  135086622     .      .     .  ID=chromosome:11;Alias=CM000673.2,chr11,NC_000...  135086621
483370    12  GRCh38  chromosome      1  133275309     .      .     .  ID=chromosome:12;Alias=CM000674.2,chr12,NC_000...  133275308
634486    13  GRCh38  chromosome      1  114364328     .      .     .  ID=chromosome:13;Alias=CM000675.2,chr13,NC_000...  114364327
In [97]: gdf.length.sum()
Out[97]: 3096629532
In [99]: chrs = [str(_) for _ in range(1, 23)] + ['X', 'Y', 'MT']
In [101]: gdf[-gdf.seqid.isin(chrs)].length.sum() / gdf.length.sum()
Out[101]: 0.0037021917421198327
</code></pre>

<p>In the above snippet first, we made a copy of <code>gdf</code> with <code>.copy()</code>. Otherwise, the original <code>gdf</code> is just a slice of <code>df</code>, and modifying it directly would result in <code>SettingWithCopyWarning</code> (see <a href="http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy" rel="noopener noreferrer" target="_blank">here</a> for more details).</p>

<p>We then calculate the length of each entry and add it back to <code>gdf</code>
 as a new column named “length”. The total length turns out to be about 
3.1 billion, and the fraction of unassembled sequences is about 0.37%.</p>

<p>Here is how the slicing works in the last two commands.</p>

<p>First, we create a list of strings that covers all seqids of well 
assembled sequences, which are all chromosomes and mitochondria. We then
 use the <code>isin</code> method to filter all entries whose seqid are in the <code>chrs</code> list.</p>

<p>A minus sign (<code>-</code>) is added to the beginning of the index to reverse the selection, because we actually want everything that is <em>not</em> in the list (i.e. we want the unassembled ones starting with KI and GL)…</p>

<p>Note: Since the assembled and unassembled sequences are distinguished
 by the type column, the last line can alternatively be rewritten as 
follows to obtain the same results.</p>

<pre><code class="language-py hljs">gdf[(gdf[<span class="hljs-string">'type'</span>] == <span class="hljs-string">'supercontig'</span>)].length.sum() / gdf.length.sum()
</code></pre>

<h4 id="how-many-genes-are-there">How Many Genes Are There?</h4>

<p>Here we focus on the entries from source ensembl, havana and 
ensembl_havana since they’re where the majority of the annotation 
entries belong.</p>

<pre><code>In [109]: edf = df[df.source.isin(['ensembl', 'havana', 'ensembl_havana'])]
In [111]: edf.sample(10)
Out[111]:
        seqid          source  type      start        end score strand phase                                         attributes
915996     16          havana   CDS   27463541   27463592     .      -     2  ID=CDS:ENSP00000457449;Parent=transcript:ENST0...
2531429     X          havana  exon   41196251   41196359     .      +     .  Parent=transcript:ENST00000462850;Name=ENSE000...
1221944    19  ensembl_havana   CDS    5641740    5641946     .      +     0  ID=CDS:ENSP00000467423;Parent=transcript:ENST0...
243070     10          havana  exon   13116267   13116340     .      +     .  Parent=transcript:ENST00000378764;Name=ENSE000...
2413583     8  ensembl_havana  exon  144359184  144359423     .      +     .  Parent=transcript:ENST00000530047;Name=ENSE000...
2160496     6          havana  exon  111322569  111322678     .      -     .  Parent=transcript:ENST00000434009;Name=ENSE000...
839952     15          havana  exon   76227713   76227897     .      -     .  Parent=transcript:ENST00000565910;Name=ENSE000...
957782     16  ensembl_havana  exon   67541653   67541782     .      +     .  Parent=transcript:ENST00000379312;Name=ENSE000...
1632979    21  ensembl_havana  exon   37840658   37840709     .      -     .  Parent=transcript:ENST00000609713;Name=ENSE000...
1953399     4          havana  exon  165464390  165464586     .      +     .  Parent=transcript:ENST00000511992;Name=ENSE000...
In [123]: edf.type.value_counts()
Out[123]:
exon                             1180596
CDS                               704604
five_prime_UTR                    142387
three_prime_UTR                   133938
transcript                         96375
gene                               42470
processed_transcript               28228
...
Name: type, dtype: int64
</code></pre>

<p>The <code>isin</code> method is used again for filtering.</p>

<p>Then, a quick value count shows that the majority of the entries are exon, coding sequence (CDS), and untranslated region (UTR).</p>

<p>These are sub-gene elements, but we are mainly looking for the gene count. As shown, there are 42,470, but we want to know more.</p>

<p>Specifically, what are their names, and what do they do? To answer 
these questions, we need to look closely at the information in the 
attributes column.</p>

<pre><code>In [127]: ndf = edf[edf.type == 'gene']
In [173]: ndf = ndf.copy()
In [133]: ndf.sample(10).attributes.values
Out[133]:
array(['ID=gene:ENSG00000228611;Name=HNF4GP1;biotype=processed_pseudogene;description=hepatocyte nuclear factor 4 gamma pseudogene 1 [Source:HGNC Symbol%3BAcc:HGNC:35417];gene_id=ENSG00000228611;havana_gene=OTTHUMG00000016986;havana_version=2;logic_name=havana;version=2',
       'ID=gene:ENSG00000177189;Name=RPS6KA3;biotype=protein_coding;description=ribosomal protein S6 kinase A3 [Source:HGNC Symbol%3BAcc:HGNC:10432];gene_id=ENSG00000177189;havana_gene=OTTHUMG00000021231;havana_version=5;logic_name=ensembl_havana_gene;version=12',
       'ID=gene:ENSG00000231748;Name=RP11-227H15.5;biotype=antisense;gene_id=ENSG00000231748;havana_gene=OTTHUMG00000018373;havana_version=1;logic_name=havana;version=1',
       'ID=gene:ENSG00000227426;Name=VN1R33P;biotype=unitary_pseudogene;description=vomeronasal 1 receptor 33 pseudogene [Source:HGNC Symbol%3BAcc:HGNC:37353];gene_id=ENSG00000227426;havana_gene=OTTHUMG00000154474;havana_version=1;logic_name=havana;version=1',
       'ID=gene:ENSG00000087250;Name=MT3;biotype=protein_coding;description=metallothionein 3 [Source:HGNC Symbol%3BAcc:HGNC:7408];gene_id=ENSG00000087250;havana_gene=OTTHUMG00000133282;havana_version=3;logic_name=ensembl_havana_gene;version=8',
       'ID=gene:ENSG00000177108;Name=ZDHHC22;biotype=protein_coding;description=zinc finger DHHC-type containing 22 [Source:HGNC Symbol%3BAcc:HGNC:20106];gene_id=ENSG00000177108;havana_gene=OTTHUMG00000171575;havana_version=3;logic_name=ensembl_havana_gene;version=5',
       'ID=gene:ENSG00000249784;Name=SCARNA22;biotype=scaRNA;description=small Cajal body-specific RNA 22 [Source:HGNC Symbol%3BAcc:HGNC:32580];gene_id=ENSG00000249784;logic_name=ncrna;version=1',
       'ID=gene:ENSG00000079101;Name=CLUL1;biotype=protein_coding;description=clusterin like 1 [Source:HGNC Symbol%3BAcc:HGNC:2096];gene_id=ENSG00000079101;havana_gene=OTTHUMG00000178252;havana_version=7;logic_name=ensembl_havana_gene;version=16',
       'ID=gene:ENSG00000229224;Name=AC105398.3;biotype=antisense;gene_id=ENSG00000229224;havana_gene=OTTHUMG00000152025;havana_version=1;logic_name=havana;version=1',
       'ID=gene:ENSG00000255552;Name=LY6G6E;biotype=protein_coding;description=lymphocyte antigen 6 complex%2C locus G6E (pseudogene) [Source:HGNC Symbol%3BAcc:HGNC:13934];gene_id=ENSG00000255552;havana_gene=OTTHUMG00000166419;havana_version=1;logic_name=ensembl_havana_gene;version=7'], dtype=object)
</code></pre>

<p>They are formatted as semicolon-separated list of tag-value pairs. 
The information we are most interested in is gene name, gene ID and 
description, and we will extract them with regular expression (regex).</p>

<pre><code class="language-py hljs"><span class="hljs-keyword">import</span> re


RE_GENE_NAME = re.compile(<span class="hljs-string">r'Name=(?P&lt;gene_name&gt;.+?);'</span>)
<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">extract_gene_name</span><span class="hljs-params">(attributes_str)</span>:</span>
    res = RE_GENE_NAME.search(attributes_str)
    <span class="hljs-keyword">return</span> res.group(<span class="hljs-string">'gene_name'</span>)


ndf[<span class="hljs-string">'gene_name'</span>] = ndf.attributes.apply(extract_gene_name)
</code></pre>

<p>First, we extract the gene names.</p>

<p>In the regex <code>Name=(?P&lt;gene_name&gt;.+?);</code> , <code>+?</code> is used instead of <code>+</code>
 because we want it to be non-greedy and let the search stop at the 
first semicolon; otherwise, the result will match up to the last 
semicolon.</p>

<p>Also, the regex is first compiled with <code>re.compile</code> instead of being used directly as in <code>re.search</code> for better performance because we will apply it to thousands of attribute strings.</p>

<p><code>extract_gene_name</code> serves as a helper function to be used in <code>pd.apply</code>, which is the method to use when a function needs to be applied on every entry of a dataframe or series.</p>

<p>In this particular case, we want to extract the gene name for every entry in <code>ndf.attributes</code>, and add the names back to <code>ndf</code> in a new column called <code>gene_name</code>.</p>

<p>Gene IDs and description are extracted in a similar way.</p>

<pre><code class="language-py hljs">RE_GENE_ID = re.compile(<span class="hljs-string">r'gene_id=(?P&lt;gene_id&gt;ENSG.+?);'</span>)
<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">extract_gene_id</span><span class="hljs-params">(attributes_str)</span>:</span>
    res = RE_GENE_ID.search(attributes_str)
    <span class="hljs-keyword">return</span> res.group(<span class="hljs-string">'gene_id'</span>)


ndf[<span class="hljs-string">'gene_id'</span>] = ndf.attributes.apply(extract_gene_id)


RE_DESC = re.compile(<span class="hljs-string">'description=(?P&lt;desc&gt;.+?);'</span>)
<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">extract_description</span><span class="hljs-params">(attributes_str)</span>:</span>
    res = RE_DESC.search(attributes_str)
    <span class="hljs-keyword">if</span> res <span class="hljs-keyword">is</span> <span class="hljs-keyword">None</span>:
        <span class="hljs-keyword">return</span> <span class="hljs-string">''</span>
    <span class="hljs-keyword">else</span>:
        <span class="hljs-keyword">return</span> res.group(<span class="hljs-string">'desc'</span>)


ndf[<span class="hljs-string">'desc'</span>] = ndf.attributes.apply(extract_description)
</code></pre>

<p>The regex for <code>RE_GENE_ID</code> is a bit more specific since we know that every <code>gene_id</code> must start with <code>ENSG</code>, where <code>ENS</code> means <a href="http://uswest.ensembl.org/index.html" rel="noopener noreferrer" target="_blank">ensembl</a> and <code>G</code> means gene.</p>

<p>For entries that don’t have any description, we will return an empty 
string. After everything is extracted, we won’t use the attributes 
column anymore, so let’s drop it to keep things nice and clean with the 
method <code>.drop</code>:</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">224</span>]: ndf.drop(<span class="hljs-string">'attributes'</span>, axis=<span class="hljs-number">1</span>, inplace=<span class="hljs-keyword">True</span>)
In [<span class="hljs-number">225</span>]: ndf.head()
Out[<span class="hljs-number">225</span>]:
   seqid          source  type  start    end score strand phase          gene_id gene_name                                               desc
<span class="hljs-number">16</span>     <span class="hljs-number">1</span>          havana  gene  <span class="hljs-number">11869</span>  <span class="hljs-number">14409</span>     .      +     .  ENSG00000223972   DDX11L1  DEAD/H-box helicase <span class="hljs-number">11</span> like <span class="hljs-number">1</span> [Source:HGNC Sym...
<span class="hljs-number">28</span>     <span class="hljs-number">1</span>          havana  gene  <span class="hljs-number">14404</span>  <span class="hljs-number">29570</span>     .      -     .  ENSG00000227232    WASH7P  WAS protein family homolog <span class="hljs-number">7</span> pseudogene [Sourc...
<span class="hljs-number">71</span>     <span class="hljs-number">1</span>          havana  gene  <span class="hljs-number">52473</span>  <span class="hljs-number">53312</span>     .      +     .  ENSG00000268020    OR4G4P  olfactory receptor family <span class="hljs-number">4</span> subfamily G member...
<span class="hljs-number">74</span>     <span class="hljs-number">1</span>          havana  gene  <span class="hljs-number">62948</span>  <span class="hljs-number">63887</span>     .      +     .  ENSG00000240361   OR4G11P  olfactory receptor family <span class="hljs-number">4</span> subfamily G member...
<span class="hljs-number">77</span>     <span class="hljs-number">1</span>  ensembl_havana  gene  <span class="hljs-number">69091</span>  <span class="hljs-number">70008</span>     .      +     .  ENSG00000186092     OR4F5  olfactory receptor family <span class="hljs-number">4</span> subfamily F member...
</code></pre>

<p>In the above call, <code>attributes</code> indicates the specific column we want to drop.</p>

<p><code>axis=1</code> means we are dropping a column instead of a row (<code>axis=0</code> by default).</p>

<p><code>inplace=True</code> means that the drop is operated on the DataFrame itself instead of returning a new copy with specified column dropped.</p>

<p>A quick <code>.head</code> look shows that the attributes column is indeed gone, and three new columns: <code>gene_name</code>, <code>gene_id</code>, and <code>desc</code> have been added.</p>

<p>Out of curiosity, let’s see if all <code>gene_id</code> and <code>gene_name</code> are unique:</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">232</span>]: ndf.shape
Out[<span class="hljs-number">232</span>]: (<span class="hljs-number">42470</span>, <span class="hljs-number">11</span>)
In [<span class="hljs-number">233</span>]: ndf.gene_id.unique().shape
Out[<span class="hljs-number">233</span>]: (<span class="hljs-number">42470</span>,)
In [<span class="hljs-number">234</span>]: ndf.gene_name.unique().shape
Out[<span class="hljs-number">234</span>]: (<span class="hljs-number">42387</span>,)
</code></pre>

<p>Surprisingly, the number of gene names is smaller than that of gene 
IDs, indicating that some gene_name must correspond to multiple gene 
IDs. Let’s find out what they are.</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">243</span>]: count_df = ndf.groupby(<span class="hljs-string">'gene_name'</span>).count().ix[:, <span class="hljs-number">0</span>].sort_values().ix[::<span class="hljs-number">-1</span>]
In [<span class="hljs-number">244</span>]: count_df.head(<span class="hljs-number">10</span>)
Out[<span class="hljs-number">244</span>]:
gene_name
SCARNA20           <span class="hljs-number">7</span>
SCARNA16           <span class="hljs-number">6</span>
SCARNA17           <span class="hljs-number">5</span>
SCARNA15           <span class="hljs-number">4</span>
SCARNA21           <span class="hljs-number">4</span>
SCARNA11           <span class="hljs-number">4</span>
Clostridiales<span class="hljs-number">-1</span>    <span class="hljs-number">3</span>
SCARNA4            <span class="hljs-number">3</span>
C1QTNF9B-AS1       <span class="hljs-number">2</span>
C11orf71           <span class="hljs-number">2</span>
Name: seqid, dtype: int64
In [<span class="hljs-number">262</span>]: count_df[count_df &gt; <span class="hljs-number">1</span>].shape
Out[<span class="hljs-number">262</span>]: (<span class="hljs-number">63</span>,)
In [<span class="hljs-number">263</span>]: count_df.shape
Out[<span class="hljs-number">263</span>]: (<span class="hljs-number">42387</span>,)
In [<span class="hljs-number">264</span>]: count_df[count_df &gt; <span class="hljs-number">1</span>].shape[<span class="hljs-number">0</span>] / count_df.shape[<span class="hljs-number">0</span>]
Out[<span class="hljs-number">264</span>]: <span class="hljs-number">0.0014863047632528842</span>
</code></pre>

<p>We group all entries by the value of <code>gene_name</code>, then count the number of items in each group with <code>.count()</code>.</p>

<p>If you inspect the output from <code>ndf.groupby('gene_name').count()</code>, all columns are counted for each group, but most of them have the same values.</p>

<p>Note that NA values won’t be considered when counting, so only take the count of the first column, <code>seqid</code> ( we use <code>.ix[:, 0]</code> to ensure that there are no NA values).</p>

<p>Then sort the count values with <code>.sort_values</code> and reverse the order with <code>.ix[::-1]</code>.</p>

<p>In the result, a gene name can be shared with up to seven gene IDs.</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">255</span>]: ndf[ndf.gene_name == <span class="hljs-string">'SCARNA20'</span>]
Out[<span class="hljs-number">255</span>]:
        seqid   source  type      start        end score strand phase          gene_id gene_name                                                                    desc
<span class="hljs-number">179399</span>   <span class="hljs-number">1</span>     ensembl  gene  <span class="hljs-number">171768070</span>  <span class="hljs-number">171768175</span>  .     +      .     ENSG00000253060  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">201037</span>   <span class="hljs-number">1</span>     ensembl  gene  <span class="hljs-number">204727991</span>  <span class="hljs-number">204728106</span>  .     +      .     ENSG00000251861  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">349203</span>   <span class="hljs-number">11</span>    ensembl  gene  <span class="hljs-number">8555016</span>    <span class="hljs-number">8555146</span>    .     +      .     ENSG00000252778  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">718520</span>   <span class="hljs-number">14</span>    ensembl  gene  <span class="hljs-number">63479272</span>   <span class="hljs-number">63479413</span>   .     +      .     ENSG00000252800  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">837233</span>   <span class="hljs-number">15</span>    ensembl  gene  <span class="hljs-number">75121536</span>   <span class="hljs-number">75121666</span>   .     -      .     ENSG00000252722  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">1039874</span>  <span class="hljs-number">17</span>    ensembl  gene  <span class="hljs-number">28018770</span>   <span class="hljs-number">28018907</span>   .     +      .     ENSG00000251818  SCARNA20  Small Cajal body specific RNA <span class="hljs-number">20</span> [Source:RFAM%<span class="hljs-number">3</span>BAcc:RF00601]
<span class="hljs-number">1108215</span>  <span class="hljs-number">17</span>    ensembl  gene  <span class="hljs-number">60231516</span>   <span class="hljs-number">60231646</span>   .     -      .     ENSG00000252577  SCARNA20  small Cajal body-specific RNA <span class="hljs-number">20</span> [Source:HGNC Symbol%<span class="hljs-number">3</span>BAcc:HGNC:<span class="hljs-number">32578</span>]
</code></pre>

<p>A closer look at all the SCARNA20 genes shows that they’re indeed all different.</p>

<p>While they share the same name, they are located at different positions of the genome.</p>

<p>Their descriptions, however, don’t seem very helpful in distinguishing them.</p>

<p>The point here is to know that gene names are not unique for all gene
 IDs, and about 0.15% of the names that are shared by multiple genes.</p>

<div class="embeddable_form-wrapper for-post" data-placement="banner" data-role="blog_subscribe" data-view="blog_subscribe#form"><form action="/blog/subscription" class="embeddable_form" data-entity="blog_subscription" data-remote="" data-view="form#form" method="post"><div class="embeddable_form-step is-email_form is-current"><div class="embeddable_form-row is-label"><div class="embeddable_form-label_title">Like what you're reading?</div><div class="embeddable_form-label">Get the latest updates first.</div></div><div class="embeddable_form-row form-field is-email_field"><input autocomplete="off" class="input is-medium" data-role="email" name="blog_subscription[email]" placeholder="Enter your email address..." type="text"><input name="blog_subscription[vertical]" value="developers" type="hidden"><input name="from_widget" value="true" type="hidden"></div><div class="embeddable_form-row is-submit"><input class="button is-green_candy is-default is-full_width" data-loader-text="Subscribing..." data-role="submit" value="Get Exclusive Updates" type="submit"></div><div class="embeddable_form-row is-privacy"><div class="embeddable_form-privacy"><div class="embeddable_form-privacy_icon"></div><div class="embeddable_form-privacy_text">No spam. Just great articles &amp; insights.</div></div></div></div><div class="embeddable_form-step is-confirmation" data-place="@blog_subscribe" data-role="confirmation"><div class="embeddable_form-row is-label is-success"><div class="embeddable_form-label_title">Like what you're reading?</div><div class="embeddable_form-label">Get the latest updates first.</div></div><div class="embeddable_form-row is-success"><div class="embeddable_form-label is-header">Thank you for subscribing!</div><div class="embeddable_form-label">Check your inbox to confirm subscription. You'll start receiving posts after you confirm.</div></div><div class="embeddable_form-row is-success"><ul class="social_share is-horizontal" data-url="https://www.toptal.com/developers/blog" data-view="layout#social_share" data-youtube-channel-url="https://www.youtube.com/channel/UCNqm_euTHZz3o5OnKhUS-oA"><li class="social_share-item is-counter" title="Total number of shares"><span class="social_share-item_num" data-role="counter_num">0</span><span class="social_share-item_text" data-role="counter_text">shares</span></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;twitter&quot;}" data-type="twitter" title="Share on Twitter"><img class="social_share-item_image" src="twitter_83c6d4.png" width="50" height="50"></a></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;facebook&quot;}" data-type="facebook" title="Share on Facebook"><img class="social_share-item_image" src="facebook_dc66c9.png" width="50" height="50"></a></li></ul></div></div></form></div><h4 id="how-long-is-a-typical-gene">How Long Is a Typical Gene?</h4>

<p>Similar to what we did when we were trying to understand the incompleteness of the genome, we can easily add a <code>length</code> column to <code>ndf</code>:</p>

<pre><code class="language-py hljs">In [<span class="hljs-number">277</span>]: ndf[<span class="hljs-string">'length'</span>] = ndf.end - ndf.start + <span class="hljs-number">1</span>
In [<span class="hljs-number">278</span>]: ndf.length.describe()
Out[<span class="hljs-number">278</span>]:
count    <span class="hljs-number">4.247000e+04</span>
mean     <span class="hljs-number">3.583348e+04</span>
std      <span class="hljs-number">9.683485e+04</span>
min      <span class="hljs-number">8.000000e+00</span>
<span class="hljs-number">25</span>%      <span class="hljs-number">8.840000e+02</span>
<span class="hljs-number">50</span>%      <span class="hljs-number">5.170500e+03</span>
<span class="hljs-number">75</span>%      <span class="hljs-number">3.055200e+04</span>
max      <span class="hljs-number">2.304997e+06</span>
Name: length, dtype: float64
</code></pre>

<p><code>.describe()</code> calculates some simple statistics based on the length values:</p>

<ul>
  <li>
    <p>Mean length of a gene is about 36,000 bases</p>
  </li>
  <li>
    <p>Median length of a gene is about 5,200 bases long</p>
  </li>
  <li>
    <p>Minimum and maximum gene lengths are about eight and 2.3 million bases long, respectively.</p>
  </li>
</ul>

<p>Because the mean is much larger than the median, it implies that 
length distribution is skewed to the right. To have a more concrete 
look, let’s plot the distribution.</p>

<pre><code class="language-py hljs"><span class="hljs-keyword">import</span> matplotlib <span class="hljs-keyword">as</span> plt


ndf.length.plot(kind=<span class="hljs-string">'hist'</span>, bins=<span class="hljs-number">50</span>, logy=<span class="hljs-keyword">True</span>)
plt.show()
</code></pre>

<p>Pandas provides a simple interface to matplotlib to make plotting very handy with DataFrames or series.</p>

<p>In this case, it says that we want a histogram plot (<code>kind='hist'</code>) with 50 bins, and let the y axis be on a log scale (<code>logy=True</code>).</p>

<p><img src="toptal-blog-image-1478601601958-60ef4533ec66c0dcd39b47dbe9b7.png" alt=""></p>

<p>From the histogram, we can see that the majority of genes are within 
the first bin. However, some gene lengths can be more than two million 
bases. Let’s find out what they are:</p>

<pre><code>In [39]: ndf[ndf.length &gt; 2e6].sort_values('length').ix[::-1]
Out[39]:
        seqid          source  type      start        end score strand phase gene_name          gene_id                                               desc   length
2309345     7  ensembl_havana  gene  146116002  148420998     .      +     .   CNTNAP2  ENSG00000174469  contactin associated protein-like 2 [Source:HG...  2304997
2422510     9  ensembl_havana  gene    8314246   10612723     .      -     .     PTPRD  ENSG00000153707  protein tyrosine phosphatase%2C receptor type ...  2298478
2527169     X  ensembl_havana  gene   31097677   33339441     .      -     .       DMD  ENSG00000198947    dystrophin [Source:HGNC Symbol%3BAcc:HGNC:2928]  2241765
440886     11  ensembl_havana  gene   83455012   85627922     .      -     .      DLG2  ENSG00000150672  discs large MAGUK scaffold protein 2 [Source:H...  2172911
2323457     8  ensembl_havana  gene    2935353    4994972     .      -     .     CSMD1  ENSG00000183117  CUB and Sushi multiple domains 1 [Source:HGNC ...  2059620
1569914    20  ensembl_havana  gene   13995369   16053197     .      +     .   MACROD2  ENSG00000172264  MACRO domain containing 2 [Source:HGNC Symbol%...  2057829
</code></pre>

<p>As you can see, the longest gene is named CNTNAP2, which is short for contactin associated protein-like 2. According to its <a href="https://en.wikipedia.org/wiki/CNTNAP2" rel="noopener noreferrer" target="_blank">wikipedia page</a>,</p>

<blockquote>
  <p>This gene encompasses almost 1.6% of chromosome 7 and is one of the largest genes in the human genome.</p>
</blockquote>

<p>Indeed! We just verified that ourselves. In contrast, what about the 
smallest genes? It turns out that they can be as short as eight bases.</p>

<pre><code>In [40]: ndf.sort_values('length').head()
Out[40]:
        seqid  source  type      start        end score strand phase   gene_name          gene_id                                               desc  length
682278     14  havana  gene   22438547   22438554     .      +     .       TRDD1  ENSG00000223997  T cell receptor delta diversity 1 [Source:HGNC...       8
682282     14  havana  gene   22439007   22439015     .      +     .       TRDD2  ENSG00000237235  T cell receptor delta diversity 2 [Source:HGNC...       9
2306836     7  havana  gene  142786213  142786224     .      +     .       TRBD1  ENSG00000282431  T cell receptor beta diversity 1 [Source:HGNC ...      12
682286     14  havana  gene   22449113   22449125     .      +     .       TRDD3  ENSG00000228985  T cell receptor delta diversity 3 [Source:HGNC...      13
1879625     4  havana  gene   10238213   10238235     .      -     .  AC006499.9  ENSG00000271544                                                         23
</code></pre>

<p>The lengths of the two extreme cases are five orders of magnitude 
apart (2.3 million vs. 8), which is enormous and which can be an 
indication of the level of diversity of life.</p>

<p>A single gene can be translated to many different proteins via a 
process called alternative splicing, something we haven’t explored. Such
 information is also inside the GFF3 file, but outside the scope of this
 post.</p>

<h4 id="gene-distribution-among-chromosomes">Gene Distribution Among Chromosomes</h4>

<p>The last thing I’d like to discuss is gene distribution among chromosomes, which also serves as an example for introducing the <code>.merge</code> method for combining two DataFrames. Intuitively, longer chromosomes likely host more genes. Let’s see if that is true.</p>

<pre><code>In [53]: ndf = ndf[ndf.seqid.isin(chrs)]
In [54]: chr_gene_counts = ndf.groupby('seqid').count().ix[:, 0].sort_values().ix[::-1]
Out[54]: chr_gene_counts
seqid
1     3902
2     2806
11    2561
19    2412
17    2280
3     2204
6     2154
12    2140
7     2106
5     2002
16    1881
X     1852
4     1751
9     1659
8     1628
10    1600
15    1476
14    1449
22     996
20     965
13     872
18     766
21     541
Y      436
Name: source, dtype: int64
</code></pre>

<p>We borrowed the <code>chrs</code> variable from the previous section,
 and used it to filter out the unassembled sequences. Based on the 
output, the largest Chromosome 1 indeed has the most genes. While 
Chromosome Y has the smallest number of genes, it is not the smallest 
chromosome.</p>

<p>Note that there seem to be no genes in the mitochondrion (MT), which is not true.</p>

<p>A bit more filtering on the first DataFrame <code>df</code> returned by <code>pd.read_csv</code> shows that all MT genes are from source insdc (which were filtered out before when generating <code>edf</code> where we only considered sources of havana, ensembl, or ensembl_havana).</p>

<pre><code>In [134]: df[(df.type == 'gene') &amp; (df.seqid == 'MT')]
Out[134]:
        seqid source  type  start    end score strand phase                                         attributes
        2514003    MT  insdc  gene    648   1601     .      +     .  ID=gene:ENSG00000211459;Name=MT-RNR1;biotype=M...
        2514009    MT  insdc  gene   1671   3229     .      +     .  ID=gene:ENSG00000210082;Name=MT-RNR2;biotype=M...
        2514016    MT  insdc  gene   3307   4262     .      +     .  ID=gene:ENSG00000198888;Name=MT-ND1;biotype=pr...
        2514029    MT  insdc  gene   4470   5511     .      +     .  ID=gene:ENSG00000198763;Name=MT-ND2;biotype=pr...
        2514048    MT  insdc  gene   5904   7445     .      +     .  ID=gene:ENSG00000198804;Name=MT-CO1;biotype=pr...
        2514058    MT  insdc  gene   7586   8269     .      +     .  ID=gene:ENSG00000198712;Name=MT-CO2;biotype=pr...
        2514065    MT  insdc  gene   8366   8572     .      +     .  ID=gene:ENSG00000228253;Name=MT-ATP8;biotype=p...
        2514069    MT  insdc  gene   8527   9207     .      +     .  ID=gene:ENSG00000198899;Name=MT-ATP6;biotype=p...
        2514073    MT  insdc  gene   9207   9990     .      +     .  ID=gene:ENSG00000198938;Name=MT-CO3;biotype=pr...
        2514080    MT  insdc  gene  10059  10404     .      +     .  ID=gene:ENSG00000198840;Name=MT-ND3;biotype=pr...
        2514087    MT  insdc  gene  10470  10766     .      +     .  ID=gene:ENSG00000212907;Name=MT-ND4L;biotype=p...
        2514091    MT  insdc  gene  10760  12137     .      +     .  ID=gene:ENSG00000198886;Name=MT-ND4;biotype=pr...
        2514104    MT  insdc  gene  12337  14148     .      +     .  ID=gene:ENSG00000198786;Name=MT-ND5;biotype=pr...
        2514108    MT  insdc  gene  14149  14673     .      -     .  ID=gene:ENSG00000198695;Name=MT-ND6;biotype=pr...
        2514115    MT  insdc  gene  14747  15887     .      +     .  ID=gene:ENSG00000198727;Name=MT-CYB;biotype=pr...
</code></pre>

<p>This example also shows how to combine two conditions during filtering with <code>&amp;</code>; the logical operator for “or” would be <code>|</code>.</p>

<p>Note that the parentheses around each condition are required, and 
this part of the syntax in Pandas is different from Python, which would 
have been literal <code>and</code> and <code>or</code>.</p>

<p>Next, let’s borrow the <code>gdf</code> DataFrame from the previous section as a source for the length of each chromosome:</p>

<pre><code>In [61]: gdf = gdf[gdf.seqid.isin(chrs)]
In [62]: gdf.drop(['start', 'end', 'score', 'strand', 'phase' ,'attributes'], axis=1, inplace=True)
In [63]: gdf.sort_values('length').ix[::-1]
Out[63]:
        seqid  source        type     length
0           1  GRCh38  chromosome  248956422
1364641     2  GRCh38  chromosome  242193529
1705855     3  GRCh38  chromosome  198295559
1864567     4  GRCh38  chromosome  190214555
1964921     5  GRCh38  chromosome  181538259
2080148     6  GRCh38  chromosome  170805979
2196981     7  GRCh38  chromosome  159345973
2514125     X  GRCh38  chromosome  156040895
2321361     8  GRCh38  chromosome  145138636
2416560     9  GRCh38  chromosome  138394717
328938     11  GRCh38  chromosome  135086622
235068     10  GRCh38  chromosome  133797422
483370     12  GRCh38  chromosome  133275309
634486     13  GRCh38  chromosome  114364328
674767     14  GRCh38  chromosome  107043718
767312     15  GRCh38  chromosome  101991189
865053     16  GRCh38  chromosome   90338345
990810     17  GRCh38  chromosome   83257441
1155977    18  GRCh38  chromosome   80373285
1559144    20  GRCh38  chromosome   64444167
1201561    19  GRCh38  chromosome   58617616
2594560     Y  GRCh38  chromosome   54106423
1647482    22  GRCh38  chromosome   50818468
1616710    21  GRCh38  chromosome   46709983
2513999    MT  GRCh38  chromosome      16569
</code></pre>

<p>The columns that are not relevant to the analysis are dropped for clarity.</p>

<p>Yes, <code>.drop</code> can also take a list of columns and drop them altogether in one operation.</p>

<p>Note that the row with a seqid of MT is still there; we will get back
 to it later. The next operation we will perform is merge the two 
datasets based on the values of seqid.</p>

<pre><code>In [73]: cdf = chr_gene_counts.to_frame(name='gene_count').reset_index()
In [75]: cdf.head(2)
Out[75]:
  seqid  gene_count
0     1        3902
1     2        2806
In [78]: merged = gdf.merge(cdf, on='seqid')
In [79]: merged
Out[79]:
   seqid  source        type     length  gene_count
0      1  GRCh38  chromosome  248956422        3902
1     10  GRCh38  chromosome  133797422        1600
2     11  GRCh38  chromosome  135086622        2561
3     12  GRCh38  chromosome  133275309        2140
4     13  GRCh38  chromosome  114364328         872
5     14  GRCh38  chromosome  107043718        1449
6     15  GRCh38  chromosome  101991189        1476
7     16  GRCh38  chromosome   90338345        1881
8     17  GRCh38  chromosome   83257441        2280
9     18  GRCh38  chromosome   80373285         766
10    19  GRCh38  chromosome   58617616        2412
11     2  GRCh38  chromosome  242193529        2806
12    20  GRCh38  chromosome   64444167         965
13    21  GRCh38  chromosome   46709983         541
14    22  GRCh38  chromosome   50818468         996
15     3  GRCh38  chromosome  198295559        2204
16     4  GRCh38  chromosome  190214555        1751
17     5  GRCh38  chromosome  181538259        2002
18     6  GRCh38  chromosome  170805979        2154
19     7  GRCh38  chromosome  159345973        2106
20     8  GRCh38  chromosome  145138636        1628
21     9  GRCh38  chromosome  138394717        1659
22     X  GRCh38  chromosome  156040895        1852
23     Y  GRCh38  chromosome   54106423         436
</code></pre>

<p>Since <code>chr_gene_counts</code> is still a Series object, which doesn’t support a merge operation, we need to convert it to a DataFrame object first with <code>.to_frame</code>.</p>

<p><code>.reset_index()</code> converts the original index (i.e. <code>seqid</code>) into a new column and resets current index as 0-based incremental numbers.</p>

<p>The output from <code>cdf.head(2)</code> shows what it looks like. Next, we used the <code>.merge</code> method to combine the two DataFrame on the seqid column (<code>on='seqid'</code>).</p>

<p>After merging <code>gdf</code> and <code>cdf</code>, the <code>MT</code> entry is still missing. This is because, by default, <code>.merge</code> operates an inner join, while left join, right join, or outer join are available by tuning the <code>how</code> parameter.</p>

<p>Please refer to the <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html#pandas.DataFrame.merge" rel="noopener noreferrer" target="_blank">documentation</a> for more details.</p>

<p>Later, you may find that there is also a related <code>.join</code> method. <code>.merge</code> and <code>.join</code> are similar yet have different APIs.</p>

<p>According to the official <a href="http://pandas.pydata.org/pandas-docs/stable/merging.html#brief-primer-on-merge-methods-relational-algebra" rel="noopener noreferrer" target="_blank">documentation</a> says</p>

<blockquote>
  <p>The related DataFrame.join method, uses merge internally for the 
index-on-index and index-on-column(s) joins, but joins on indexes by 
default rather than trying to join on common columns (the default 
behavior for merge). If you are joining on index, you may wish to use 
DataFrame.join to save yourself some typing.</p>
</blockquote>

<p>Basically, <code>.merge</code> is more general-purpose and used by <code>.join</code>.</p>

<p>Finally, we are ready to calculate the correlation between chromosome <code>length</code> and <code>gene_count</code>.</p>

<pre><code>In [81]: merged[['length', 'gene_count']].corr()
Out[81]:
              length  gene_count
length      1.000000    0.728221
gene_count  0.728221    1.000000
</code></pre>

<p>By default <code>.corr</code> calculates the <a href="https://en.wikipedia.org/wiki/Pearson_product-moment_correlation_coefficient" rel="noopener noreferrer" target="_blank">Pearson correlation</a> between all pairs of columns in a dataframe.</p>

<p>But we have only a single pair of columns in this case, and the correlation turns out to be positive – 0.73.</p>

<p>In other words, the larger the chromosome, the more likely it is to 
have more genes. Let’s also plot the two columns after sorting the value
 pairs by <code>length</code>:</p>

<pre><code class="language-py hljs">ax = merged[[<span class="hljs-string">'length'</span>, <span class="hljs-string">'gene_count'</span>]].sort_values(<span class="hljs-string">'length'</span>).plot(x=<span class="hljs-string">'length'</span>, y=<span class="hljs-string">'gene_count'</span>, style=<span class="hljs-string">'o-'</span>)
<span class="hljs-comment"># add some margin to both ends of x axis</span>
xlim = ax.get_xlim()
margin = xlim[<span class="hljs-number">0</span>] * <span class="hljs-number">0.1</span>
ax.set_xlim([xlim[<span class="hljs-number">0</span>] - margin, xlim[<span class="hljs-number">1</span>] + margin])
<span class="hljs-comment"># Label each point on the graph</span>
<span class="hljs-keyword">for</span> (s, x, y) <span class="hljs-keyword">in</span> merged[[<span class="hljs-string">'seqid'</span>, <span class="hljs-string">'length'</span>, <span class="hljs-string">'gene_count'</span>]].sort_values(<span class="hljs-string">'length'</span>).values:
    ax.text(x, y - <span class="hljs-number">100</span>, str(s))
</code></pre>

<p><img src="toptal-blog-image-1478601215377-37d251d5cf8742c53a16f90574c1.png" alt=""></p>

<p>As seen in image above, even though it is a positive correlation 
overall, it does not hold for all chromosomes. In particular, for 
Chromosome 17, 16, 15, 14, 13, the correlation is actually negative, 
meaning the number of genes on the chromosome decreases as the 
chromosome size increases.</p>

<h2 id="findings-and-future-research">Findings and Future Research</h2>

<p>That ends our tutorial on the manipulation of an annotation file for 
human genome in GFF3 format with the SciPy stack. The tools we’ve mainly
 used include IPython, Pandas, and matplotlib. During the tutorial, not 
only have we learned some of the most common and useful operations in 
Pandas, we also answered some very interesting questions about our 
genome. In summary:</p>

<ol>
  <li>About 0.37% of the human genome is still incomplete even though the first draft came out over 15 years ago.</li>
  <li>There are about 42,000 genes in the human genome based on this particular <a href="ftp://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/Homo_sapiens.GRCh38.85.gff3.gz" rel="noopener noreferrer" target="_blank">GFF3 file</a> we used.</li>
  <li>The length of a gene can range from a few dozen to over two million bases.</li>
  <li>Genes are not evenly distributed among the chromosomes. Overall, 
the larger the chromosome, the more genes it hosts, but for a subset of 
the chromosomes, the correlation can be negative.</li>
</ol>

<p>The GFF3 file is very rich in annotation information, and we have 
just scratched the surface. If you are interested in further 
exploration, here are a few questions you can play with:</p>

<ol>
  <li>How many transcripts does a gene typically have? What percentage of genes have more than 1 transcript?</li>
  <li>How many isoforms does a transcript typically have?</li>
  <li>How many exons, CDS, and UTRs does a transcript typically have? What sizes are they?</li>
  <li>Is it possible to categorize the genes based on their function as described in the description column?</li>
</ol>
<div class="blog_author_big_banner is-active" data-ga-action="ClickedOnHireTheAuthor_bottom" data-ga-label="comprehensive-introduction-your-genome-scipy" data-view="content#bottom_banner"><h2>About the author</h2><div class="skill_talent for-blog_post"><div class="skill_talent-extra"><a class="skill_talent-photo_wrapper" href="https://www.toptal.com/resume/zhuyi-xue"><img alt="Zhuyi Xue" class="skill_talent-photo" src="large_45dcc56c2c3475b8c81e30a9ae7812c2.jpg"><div class="skill_talent-photo_hint">View full profile »</div></a><div class="blog_author_big_banner-cta"><a class="button is-green_candy is-default is-full_width is-fixed skill_talent-hire" data-role="banner_link" href="https://www.toptal.com/hire?interested_in=developers&amp;skill=python">Hire the Author</a></div></div><div class="skill_talent-main"><a class="link skill_talent-name" href="https://www.toptal.com/resume/zhuyi-xue">Zhuyi Xue, Canada</a><div class="clearfix"><div class="skill_talent-member_since">member since March 5, 2015</div></div><div class="skill_talent-skills" data-view="skill_talent#tags" style="max-height: none;"><a class="tag skill_talent-skill" data-role="tag" href="https://www.toptal.com/machine-learning">Machine Learning</a><a class="tag skill_talent-skill" data-role="tag" href="https://www.toptal.com/google-app-engine">Google App Engine</a><a class="tag skill_talent-skill" data-role="tag" href="https://www.toptal.com/python">Pandas</a><a class="tag" data-role="more" href="https://www.toptal.com/resume/zhuyi-xue" style="">+<span data-role="count">9</span> more</a></div><div class="skill_talent-description">Zhuyi
 is a skilled Python developer with over seven years of experience. He 
is also proficient in JavaScript and Scala. He has collected experience 
in a wide array of technologies with strong focuses on PyData stack, 
Django, and cloud computing. He is very detail-oriented and proactive, 
with great communication skills. <a class="link" href="https://www.toptal.com/resume/zhuyi-xue">[click to continue...]</a></div></div></div></div><div class="post-join"><a class="button is-white_and_green is-default is-full_width" href="https://www.toptal.com/python">Hiring? Meet the Top 10 Freelance Python Developers for Hire in April 2019</a></div><div class="template" data-role="middle_content_template" data-template="{&quot;html&quot;:&quot;\u003cdiv class=\&quot;embeddable_form-wrapper for-post\&quot; data-placement=\&quot;banner\&quot; data-role=\&quot;blog_subscribe\&quot; data-view=\&quot;blog_subscribe#form\&quot;\u003e\u003cform action=\&quot;/blog/subscription\&quot; class=\&quot;embeddable_form\&quot; data-entity=\&quot;blog_subscription\&quot; data-remote=\&quot;\&quot; data-view=\&quot;form#form\&quot; method=\&quot;post\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-step is-email_form is-current\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-row is-label\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-label_title\&quot;\u003eLike what you\u0026#39;re reading?\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-label\&quot;\u003eGet the latest updates first.\u003c/div\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-row form-field is-email_field\&quot;\u003e\u003cinput autocomplete=\&quot;off\&quot; class=\&quot;input is-medium\&quot; data-role=\&quot;email\&quot; name=\&quot;blog_subscription[email]\&quot; placeholder=\&quot;Enter your email address...\&quot; type=\&quot;text\&quot; /\u003e\u003cinput name=\&quot;blog_subscription[vertical]\&quot; type=\&quot;hidden\&quot; value=\&quot;developers\&quot; /\u003e\u003cinput name=\&quot;from_widget\&quot; type=\&quot;hidden\&quot; value=\&quot;true\&quot; /\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-row is-submit\&quot;\u003e\u003cinput class=\&quot;button is-green_candy is-default is-full_width\&quot; data-loader-text=\&quot;Subscribing...\&quot; data-role=\&quot;submit\&quot; type=\&quot;submit\&quot; value=\&quot;Get Exclusive Updates\&quot; /\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-row is-privacy\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-privacy\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-privacy_icon\&quot;\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-privacy_text\&quot;\u003eNo spam. Just great articles \u0026amp; insights.\u003c/div\u003e\u003c/div\u003e\u003c/div\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-step is-confirmation\&quot; data-place=\&quot;@blog_subscribe\&quot; data-role=\&quot;confirmation\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-row is-label is-success\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-label_title\&quot;\u003eLike what you\u0026#39;re reading?\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-label\&quot;\u003eGet the latest updates first.\u003c/div\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-row is-success\&quot;\u003e\u003cdiv class=\&quot;embeddable_form-label is-header\&quot;\u003eThank you for subscribing!\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-label\&quot;\u003eCheck your inbox to confirm subscription. You'll start receiving posts after you confirm.\u003c/div\u003e\u003c/div\u003e\u003cdiv class=\&quot;embeddable_form-row is-success\&quot;\u003e\u003cul class=\&quot;social_share is-horizontal\&quot; data-url=\&quot;https://www.toptal.com/developers/blog\&quot; data-view=\&quot;layout#social_share\&quot; data-youtube-channel-url=\&quot;https://www.youtube.com/channel/UCNqm_euTHZz3o5OnKhUS-oA\&quot;\u003e\u003cli class=\&quot;social_share-item is-counter\&quot; title=\&quot;Total number of shares\&quot;\u003e\u003cspan class=\&quot;social_share-item_num\&quot; data-role=\&quot;counter_num\&quot;\u003e0\u003c/span\u003e\u003cspan class=\&quot;social_share-item_text\&quot; data-role=\&quot;counter_text\&quot;\u003eshares\u003c/span\u003e\u003c/li\u003e\u003cli class=\&quot;social_share-item\&quot;\u003e\u003ca class=\&quot;social_share-item_link\&quot; data-role=\&quot;link\&quot; data-track-click=\&quot;SocialShareButtonClicked\&quot; data-track-default-data=\&quot;{\u0026quot;network\u0026quot;:\u0026quot;twitter\u0026quot;}\&quot; data-type=\&quot;twitter\&quot; title=\&quot;Share on Twitter\&quot;\u003e\u003cimg class=\&quot;social_share-item_image\&quot; height=\&quot;50\&quot; src=\&quot;https://assets.toptal.io/assets/front/static/public/primitives/social/share_bar/twitter_83c6d4.png\&quot; width=\&quot;50\&quot; /\u003e\u003c/a\u003e\u003c/li\u003e\u003cli class=\&quot;social_share-item\&quot;\u003e\u003ca class=\&quot;social_share-item_link\&quot; data-role=\&quot;link\&quot; data-track-click=\&quot;SocialShareButtonClicked\&quot; data-track-default-data=\&quot;{\u0026quot;network\u0026quot;:\u0026quot;facebook\u0026quot;}\&quot; data-type=\&quot;facebook\&quot; title=\&quot;Share on Facebook\&quot;\u003e\u003cimg class=\&quot;social_share-item_image\&quot; height=\&quot;50\&quot; src=\&quot;https://assets.toptal.io/assets/front/static/public/primitives/social/share_bar/facebook_dc66c9.png\&quot; width=\&quot;50\&quot; /\u003e\u003c/a\u003e\u003c/li\u003e\u003c/ul\u003e\u003c/div\u003e\u003c/div\u003e\u003c/form\u003e\u003c/div\u003e&quot;}"></div></div><div class="blog-comments"><div id="disqus_thread_container"><div id="disqus_thread"><iframe id="dsq-app151" name="dsq-app151" allowtransparency="true" scrolling="no" tabindex="0" title="Disqus" style="width: 1px ! important; min-width: 100% ! important; border: medium none ! important; overflow: hidden ! important; height: 0px ! important;" src="https://disqus.com/embed/comments/?base=default&amp;f=toptal&amp;t_i=python%2Fcomprehensive-introduction-your-genome-scipy&amp;t_u=https%3A%2F%2Fwww.toptal.com%2Fpython%2Fcomprehensive-introduction-your-genome-scipy&amp;t_d=Explore%20the%20Human%20Genome%20with%20the%20SciPy%20Stack%20%7C%20Toptal&amp;t_t=Explore%20the%20Human%20Genome%20with%20the%20SciPy%20Stack%20%7C%20Toptal&amp;s_o=default&amp;d_m=2#version=85f2151add92f574efd2c4ee2f7b9f67" width="100%" frameborder="0"></iframe></div></div><script>function isIos() {
  return /(iPad|iPhone|iPod)/g.test(navigator.userAgent);
}

function adjustHeightInIos() {
  if (isIos()) {
    var height = $('#disqus_thread').outerHeight();
    $('#disqus_thread_container').height(height);
  }
}

if (isIos()) {
  var el = document.getElementById('disqus_thread_container');
  el.style.height = '1px';
  el.style.overflow = 'hidden';
  el.style.width = '100%';
}

function disqus_config() {
  this.callbacks.onNewComment = [function() {
    $.post('/blog/comments/on_create');
    adjustHeightInIos();
  }];

  this.callbacks.onReady = [adjustHeightInIos];
}

var disqus_shortname = 'toptal';
var disqus_identifier = 'python/comprehensive-introduction-your-genome-scipy';

(function() {
  var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
  dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
  (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
})();</script><noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript><a class="dsq-brlink" href="https://disqus.com/" rel="noopener noreferrer">comments powered by <span class="logo-disqus">Disqus</span></a></div></div></div><div class="two_columns_layout-extra"><div class="blog-block-wrapper"><div class="blog-block-header">Subscribe</div><div class="blog-block"><div class="blog-block-row"><div class="embeddable_form-wrapper is-horizontal" data-placement="sidebar" data-role="blog_subscribe" data-view="blog_subscribe#form"><form action="/blog/subscription" class="embeddable_form" data-entity="blog_subscription" data-remote="" data-view="form#form" method="post"><div class="embeddable_form-step is-email_form is-current"><div class="embeddable_form-row is-label"><div class="embeddable_form-label_title">Free email updates</div><div class="embeddable_form-label">Get the latest content first.</div></div><div class="embeddable_form-row form-field is-email_field"><input autocomplete="off" class="input is-medium" data-role="email" name="blog_subscription[email]" placeholder="Enter your email address..." type="text"><input name="blog_subscription[vertical]" value="developers" type="hidden"><input name="from_widget" value="true" type="hidden"></div><div class="embeddable_form-row is-submit"><input class="button is-green_candy is-default is-full_width" data-loader-text="Subscribing..." data-role="submit" value="Get Exclusive Updates" type="submit"></div><div class="embeddable_form-row is-privacy"><div class="embeddable_form-privacy"><div class="embeddable_form-privacy_icon"></div><div class="embeddable_form-privacy_text">No spam. Just great articles &amp; insights.</div></div></div></div><div class="embeddable_form-step is-confirmation" data-place="@blog_subscribe" data-role="confirmation"><div class="embeddable_form-row is-label is-success"><div class="embeddable_form-label_title">Free email updates</div><div class="embeddable_form-label">Get the latest content first.</div></div><div class="embeddable_form-row is-success"><div class="embeddable_form-label is-header">Thank you for subscribing!</div><div class="embeddable_form-label">Check your inbox to confirm subscription. You'll start receiving posts after you confirm.</div></div><div class="embeddable_form-row is-success"><ul class="social_share is-horizontal" data-url="https://www.toptal.com/developers/blog" data-view="layout#social_share" data-youtube-channel-url="https://www.youtube.com/channel/UCNqm_euTHZz3o5OnKhUS-oA"><li class="social_share-item is-counter" title="Total number of shares"><span class="social_share-item_num" data-role="counter_num">0</span><span class="social_share-item_text" data-role="counter_text">shares</span></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;twitter&quot;}" data-type="twitter" title="Share on Twitter" rel="noopener noreferrer"><img class="social_share-item_image" src="twitter_83c6d4.png" width="50" height="50"></a></li><li class="social_share-item"><a class="social_share-item_link" data-role="link" data-track-click="SocialShareButtonClicked" data-track-default-data="{&quot;network&quot;:&quot;facebook&quot;}" data-type="facebook" title="Share on Facebook" rel="noopener noreferrer"><img class="social_share-item_image" src="facebook_dc66c9.png" width="50" height="50"></a></li></ul></div></div></form></div></div></div><div class="blog-block-header">Trending articles</div><div class="blog-block"><nav class="blog-trending"><a class="post is-preview blog-trending-item" href="https://www.toptal.com/flutter/flutter-tutorial"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121861/side_list_cover-0408-FlutterMessangerDemo-Luke_Newsletter-30d5a65064b44f0ef56a801d4811964a.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121861/side_list_cover-0408-FlutterMessangerDemo-Luke_Newsletter-30d5a65064b44f0ef56a801d4811964a.png"></span></span><span class="post-preview"><span class="post-title">Flutter Tutorial: How to Create Your First Flutter App</span><span class="post_meta"><span class="post_meta-date">2 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/angular/angular-components-overview-101"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121860/side_list_cover-0401-AngularComponents-Waldek_Newsletter-d5c55b4c7e2ac8887ee93b53a6e4d9cb.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121860/side_list_cover-0401-AngularComponents-Waldek_Newsletter-d5c55b4c7e2ac8887ee93b53a6e4d9cb.png"></span></span><span class="post-preview"><span class="post-title">Angular Components 101 — an Overview</span><span class="post_meta"><span class="post_meta-date">2 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/devops/scaling-microservices-applications"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121857/side_list_cover-0401-HowToMathematicallyConfigureYourApplicationForAuto-Scaling-Waldek_Newsletter-003822358e45fdefb90d9dc0354cc33f.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121857/side_list_cover-0401-HowToMathematicallyConfigureYourApplicationForAuto-Scaling-Waldek_Newsletter-003822358e45fdefb90d9dc0354cc33f.png"></span></span><span class="post-preview"><span class="post-title">Do the Math: Scaling Microservices Applications with Orchestrators</span><span class="post_meta"><span class="post_meta-date">7 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/wordpress/wordpress-roots-stack"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121855/side_list_cover-0104-WodpressRoots-Luke_Social-14cea8a367cb03c4a39a3860f9a11823.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121855/side_list_cover-0104-WodpressRoots-Luke_Social-14cea8a367cb03c4a39a3860f9a11823.png"></span></span><span class="post-preview"><span class="post-title">Modern WordPress Development Workflow with the Roots Stack</span><span class="post_meta"><span class="post_meta-date">9 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/bitcoin/intro-to-bitcoin-lightning-network"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121851/side_list_cover-0327-IntroToBitcoinLightning-Waldek_Banner-e63121da140c50e3b77db0e134d4f9f8.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121851/side_list_cover-0327-IntroToBitcoinLightning-Waldek_Banner-e63121da140c50e3b77db0e134d4f9f8.png"></span></span><span class="post-preview"><span class="post-title">Scale with Speed: The Bitcoin Lightning Network Explained</span><span class="post_meta"><span class="post_meta-date">10 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/wordpress/wordpress-maintenance-tips"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121848/side_list_cover-0326-WordpressMaintenanceTips-Luke_Newsletter-24d38612919504bbefbe7d83c73c42c9.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121848/side_list_cover-0326-WordpressMaintenanceTips-Luke_Newsletter-24d38612919504bbefbe7d83c73c42c9.png"></span></span><span class="post-preview"><span class="post-title">10 Tips to Make WordPress Maintenance Smooth</span><span class="post_meta"><span class="post_meta-date">16 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/software/life-critical-systems"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121844/side_list_cover-0319-LifeCriticalSystems-Luke_Newsletter-8c3a1ee7ec272731e84d07c1e1cb92e8.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121844/side_list_cover-0319-LifeCriticalSystems-Luke_Newsletter-8c3a1ee7ec272731e84d07c1e1cb92e8.png"></span></span><span class="post-preview"><span class="post-title">Innovation with Life-critical Systems</span><span class="post_meta"><span class="post_meta-date">23 days ago</span></span></span></a><a class="post is-preview blog-trending-item" href="https://www.toptal.com/python/python-parameterized-design-patterns"><span class="post_cover is-preview"><noscript><img class="post_cover-image" src="https://uploads.toptal.io/blog/cover_image/121843/side_list_cover-0318-PhytonParameterization_Dan_Newsletter-b9dfeef97a76cce4ba5d826d3fff10ce.png" /></noscript><span class="post_cover-image" data-role="deferred" data-src="https://uploads.toptal.io/blog/cover_image/121843/side_list_cover-0318-PhytonParameterization_Dan_Newsletter-b9dfeef97a76cce4ba5d826d3fff10ce.png"></span></span><span class="post-preview"><span class="post-title">Ensuring Clean Code: A Look at Python, Parameterized</span><span class="post_meta"><span class="post_meta-date">23 days ago</span></span></span></a></nav></div><div class="blog-block-header">Relevant Technologies</div><div class="blog-block"><div class="blog-block-row"><ul class="blog-relevant"><li class="blog-relevant-item"><a class="link is-blue" href="https://www.toptal.com/python">Python</a></li><li class="blog-relevant-item"><a class="link is-blue" href="https://www.toptal.com/data-science">Data Science</a></li></ul></div></div><div class="blog-block has-no_background blog_banner for-author blog_author_banner is-active" data-ga-action="ClickedOnHireTheAuthor_lateral" data-ga-label="comprehensive-introduction-your-genome-scipy" data-view="content#sidebar_banner"><div class="blog-block-header">About the author</div><div class="blog_author_banner-photo_wrapper"><img alt="Zhuyi Xue" class="blog_author_banner-photo" src="large_45dcc56c2c3475b8c81e30a9ae7812c2.jpg"></div><div class="blog_author_banner-content" data-role="author_content" style="height: 180px;"><div class="blog_author_banner-name"><a href="https://www.toptal.com/resume/zhuyi-xue">Zhuyi Xue</a></div><div class="blog_author_banner-role">Python Developer</div><div class="blog_author_banner-about" data-role="author_about">Zhuyi
 is a skilled Python developer with over seven years of experience. He 
is also proficient in JavaScript and Scala. He has collected experience 
in a wide array of technologies with strong focuses on PyData stack, 
Django, and cloud computing. He is very detail-oriented and proactive, 
with great communication skills.</div><div class="blog_author_banner-content_footer"></div></div><div class="blog_author_banner-cta"><a class="button is-green_candy is-default is-full_width" data-role="banner_link" href="https://www.toptal.com/hire?interested_in=developers&amp;skill=python">Hire the Author</a></div></div></div></div></div></section></main><footer class="hide_in_webview"><section class="grid-row has-solid_blue_bg page_footer_map-wrapper"><div class="grid-row-inner has-huge_padding page_footer_map"><div class="page_footer_map-content">Toptal connects the&nbsp;<a href="https://www.toptal.com/top-3-percent">top&nbsp;3%</a> of&nbsp;freelance talent all over&nbsp;the&nbsp;world.</div></div></section><section class="grid-row page_footer_sample_profile-wrapper"><div class="grid-row-inner has-big_padding"><h2 class="header is-row">Toptal Developers</h2><div class="skills_list"><ul class="multi_column_list qa-skills_list"><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/android">Android Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/angular-js">AngularJS Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/back-end">Back-End Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/c-plus-plus">C++ Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/data-science">Data Scientists</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/devops">DevOps Engineers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/emberjs">Ember.js Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/freelance">Freelance Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/front-end">Front-End Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/full-stack">Full-Stack Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/html5">HTML5 Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/ios">iOS Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/java">Java Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/javascript">JavaScript Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/machine-learning">Machine Learning Engineers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/magento">Magento Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/app">Mobile App Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/dot-net">.NET Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/nodejs">Node.js Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/php">PHP Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/python">Python Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/react">React.js Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/ruby">Ruby Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/ruby-on-rails">Ruby on Rails Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/salesforce">Salesforce Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/scala">Scala Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/software">Software Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/unity-unity3d">Unity or Unity3D Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/virtual-reality">Virtual Reality Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/web">Web Developers</a></li><li class="multi_column_list-item"><a class="multi_column_list-link" href="https://www.toptal.com/wordpress">WordPress Developers</a></li></ul><div class="skills_list-see_all_wrapper"><a class="button is-tiny is-white is-uppercased" href="https://www.toptal.com/developers"><div class="skills_list-see_all_icon for-developers"></div>See more freelance developers</a></div></div></div></section><div class="skills_enterprise_link"><a class="skills_enterprise_link-url" href="https://www.toptal.com/enterprise">Learn how enterprises benefit from Toptal experts.</a></div><section class="grid-row has-top_border page_footer_join-wrapper" id=""><div class="grid-row-inner has-padding page_footer_join"><h2 class="page_footer_join-header">Join the Toptal community.</h2><div class="page_footer_join-spacer"></div><nav class="page_footer_join-actions"><div class="page_footer_join-item"><a class="button is-green_candy is-big is-fixed page_footer_join-button" href="https://www.toptal.com/hire?interested_in=developers&amp;skill=python">Hire a developer</a></div><div class="page_footer_join-item is-or">or</div><div class="page_footer_join-item"><a class="button is-green_candy is-big is-fixed page_footer_join-button" href="https://www.toptal.com/developers/join">Apply as a Developer</a></div></nav></div></section><section class="grid-row page_footer_links-wrapper"><div class="grid-row-inner is-wide has-big_padding"><nav class="page_footer_links"><section class="page_footer_links-column is-1"><h3 class="page_footer_links-column_header">Highest In-Demand Talent</h3><ul class="page_footer_links-list"><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/ios">iOS Developers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/front-end">Front-End Developers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/designers/ux">UX Designers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/designers/ui">UI Designers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/finance/financial-modeling">Financial Modeling Consultants</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/finance/interim-cfos">Interim CFOs</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/project-managers/digital">Digital Project Managers</a></li></ul></section><section class="page_footer_links-column is-2"><h3 class="page_footer_links-column_header">About</h3><ul class="page_footer_links-list"><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/top-3-percent">Top 3%</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/clients">Clients</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/developers">Freelance Developers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/designers">Freelance Designers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/finance">Freelance Finance Experts</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/project-managers">Freelance Project Managers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/product-managers">Freelance Product Managers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/about">About Us</a></li></ul></section><section class="page_footer_links-column is-3"><h3 class="page_footer_links-column_header">Contact</h3><ul class="page_footer_links-list"><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/contact">Contact Us</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/press-center">Press Center</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/careers">Careers</a></li><li class="page_footer_links-item"><a class="page_footer_links-link" href="https://www.toptal.com/faq">FAQ</a></li></ul></section><section class="page_footer_links-column is-4"><h3 class="page_footer_links-column_header">Social</h3><ul class="page_footer_links-list"><li class="page_footer_links-item for-social_link"><a class="page_footer_links-social_link is-facebook" href="https://www.facebook.com/toptal" rel="noopener noreferrer">Facebook</a></li><li class="page_footer_links-item for-social_link"><a class="page_footer_links-social_link is-twitter" href="https://twitter.com/toptal" rel="noopener noreferrer">Twitter</a></li><li class="page_footer_links-item for-social_link"><a class="page_footer_links-social_link is-instagram" href="https://www.instagram.com/toptal" rel="noopener noreferrer">Instagram</a></li><li class="page_footer_links-item for-social_link"><a class="page_footer_links-social_link is-linkedin" href="https://www.linkedin.com/company/toptal" rel="noopener noreferrer">LinkedIn</a></li></ul></section></nav></div></section><section class="grid-row page_footer_legal-wrapper"><div class="grid-row-inner has-medium_padding page_footer_legal is-wide"><div class="logo-wrapper page_footer_legal-logotype"><a class="logo is-mini is-link" href="https://www.toptal.com/">Toptal</a><p class="logo-motto is-mini">Hire the top 3% of freelance talent</p></div><ul class="page_footer_legal-links"><li class="page_footer_legal-links_item is-copyright"><p class="page_footer_legal-copyright">© Copyright 2010 - 2019 Toptal, LLC</p></li><li class="page_footer_legal-links_item has-right_margin"><a class="page_footer_legal-link" href="https://www.toptal.com/privacy">Privacy Policy</a></li><li class="page_footer_legal-links_item"><a class="page_footer_legal-link" href="https://www.toptal.com/tos">Website Terms</a></li></ul></div></section></footer></div><div class="layout_layer-hover_cover"></div><div class="layout-overlay" data-view="layout#overlay"></div><div class="notification-container" data-view="notifications#service"></div><script>setTimeout(function () {
  classNameWithFont = document.documentElement.className + ' wf-active'
  document.documentElement.className = classNameWithFont
}, 0);</script><script src="public-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><script src="blog-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><script src="cookie_banner-65b599e9a9ece62d8bb8513c92224c490fe27e50.js"></script><div class="layout-counters"><script async="true" src="js"></script><script>if (typeof(window.Widgets) !== 'undefined') {
  var trackingId = 'UA-21104039-1'

  if (trackingId.length && !(window.googleAnalytics instanceof Widgets.GoogleAnalytics)) {
    window.googleAnalytics = new Widgets.GoogleAnalytics(
      trackingId,
      '',
      '',
      false
    )
  }
}</script><script async="" defer="defer" src="A6qhozMiYLpV2VKc8wSHaEmc3saOMhcrxAFcCFAt" type="text/javascript"></script><noscript><img src="https://cdn.distiltag.com/api/v1/noscript/A6qhozMiYLpV2VKc8wSHaEmc3saOMhcrxAFcCFAt?GA_TID=UA-21104039-1&amp;amp;GA_CDI=14" /></noscript><script>// We are checking if window.analytics has the track function in case some dom element has the id equal analytics
if (!window.analytics || typeof window.analytics.track !== 'function') {
  // We do this because if some dom element has id equal analytics we need to override it, otherwise it wouldn't set analytics right
  window.analytics = undefined
  !function(){var analytics=window.analytics=window.analytics||[];if(!analytics.initialize)if(analytics.invoked)window.console&&console.error&&console.error("Segment snippet included twice.");else{analytics.invoked=!0;analytics.methods=["setAnonymousId", "trackSubmit","trackClick","trackLink","trackForm","pageview","identify","group","track","ready","alias","page","once","off","on"];analytics.factory=function(t){return function(){var e=Array.prototype.slice.call(arguments);e.unshift(t);analytics.push(e);return analytics}};for(var t=0;t<analytics.methods.length;t++){var e=analytics.methods[t];analytics[e]=analytics.factory(e)}analytics.load=function(t){var e=document.createElement("script");e.type="text/javascript";e.async=!0;e.src=("https:"===document.location.protocol?"https://":"http://")+"cdn.segment.com/analytics.js/v1/"+t+"/analytics.min.js";var n=document.getElementsByTagName("script")[0];n.parentNode.insertBefore(e,n)};analytics.SNIPPET_VERSION="3.0.1";
  analytics.load('jnS4QsHOCAOeG6XvMDCjD9n9bAcQ53Mb');
  }}();
}
(function() {
  function readVisitorIdCookie() {
    var cookieName = 'visitor_id=';
    var cookies = document.cookie.split(';');
    for (var i = 0; i < cookies.length; i++) {
      var cookie = cookies[i].replace(/^\s+|\s+$/g, '');
      if (cookie.indexOf(cookieName) === 0) {
        return cookie.split('=')[1];
      }
    }
  }
  analytics.ready(function() { window.analytics['isReady'] = true });
  analytics.setAnonymousId(readVisitorIdCookie());
  
  analytics.page('posts#show');
})();</script><script async="" defer="defer" id="hs-script-loader" src="2799924.js" type="text/javascript"></script><script>var _hsq = window._hsq = window._hsq || [];
</script><script>!function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod?
n.callMethod.apply(n,arguments):n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');

fbq('init', '463369723801939');
fbq('track', 'ViewContent');</script><noscript><img height="1" src="https://www.facebook.com/tr?id=463369723801939&amp;ev=ViewContent&amp;noscript=1" width="1" /></noscript><iframe scrolling="no" src="https://www.attributiontracker.com/tracker/account_visit?tt_visit=1" width="1" frameborder="0" height="1"></iframe><script>(function(A, d, S, t, a, g, e) {
  A['AdstageAnalytics'] = a;
  A[a] = A[a] || function() {
    (A[a].q = A[a].q || []).push(arguments)
  };
  A[a].l = 1 * new Date();
  g = d.createElement(S), e = d.getElementsByTagName(S)[0];
  g.async = 1;
  g.src = t;
  e.parentNode.insertBefore(g, e);
})(window, document, 'script', '//assets.adstage.io/analytics.js', '_as');
_as('create', 'e2dd65eb-07a1-4d93-874a-fe6dac5a8b64');
_as('send', 'pageview');</script><script>_bizo_data_partner_id = "8663";
(function() {
  var s = document.getElementsByTagName("script")[0];
  var b = document.createElement("script");
  b.type = "text/javascript";
  b.async = true;
  b.src = (window.location.protocol === "https:" ? "https://sjs" : "http://js") + ".bizographics.com/insight.min.js";
  s.parentNode.insertBefore(b, s);
})();</script><noscript><img alt="" height="1" src="//www.bizographics.com/collect/?pid=8663&amp;fmt=gif" style="display:none;" width="1" /></noscript></div><div id="breadcrumbs" class="breadcrumbs"><span itemscope="itemscope" itemtype="http://data-vocabulary.org/Breadcrumb"><a itemprop="url" href="https://www.toptal.com/"><span itemprop="title">Home</span></a></span> › <span itemscope="itemscope" itemtype="http://data-vocabulary.org/Breadcrumb"><a itemprop="url" href="https://www.toptal.com/developers/blog"><span itemprop="title">Blog</span></a></span> › <span itemscope="itemscope" itemtype="http://data-vocabulary.org/Breadcrumb"><a class="current" itemprop="url" href="https://www.toptal.com/python/comprehensive-introduction-your-genome-scipy"><span itemprop="title">A Comprehensive Introduction To Your Genome With the SciPy Stack</span></a></span></div><nav class="blog-simple-header grid-row page_header_menu-wrapper is-fixed" data-role="sticky_menu_wrapper" style=""><div class="grid-row-inner is-wide page_header_menu has-padding"><div class="page_header_top-logotype_wrapper"><div class="page_header_top-logotype"><a class="logo is-big is-link" href="https://www.toptal.com/developers"></a></div><div class="page_header_top-main_title"><b>Hiring?</b> Toptal handpicks <b><a href="https://www.toptal.com/python" onclick='window.googleAnalytics.trackEvent(event, "BlogCTAs", "ClickedOnHeaderHeadline", "ClickedOnHeaderHeadline")'>top Python engineers</a></b> to suit your needs.</div></div><nav class="page_header_top-nav_wrapper"><ul class="page_header_top-nav" data-role="menu"><li class="page_header_top-nav_item"><a class="button is-green_candy is-small page_header_top-nav_link is-button" data-role="companies_apply_button" href="https://www.toptal.com/python" onclick='window.googleAnalytics.trackEvent(event, "BlogCTAs", "ClickedOnHeaderCTA", "ClickedOnHeaderCTA");'>Start hiring</a></li><li class="page_header_top-nav_item"><a class="button is-small is-light page_header_top-nav_link is-button" href="https://www.toptal.com/users/login">Log in</a></li></ul></nav></div></nav><iframe style="display: none;"></iframe><iframe scrolling="no" allowtransparency="true" src="https://platform.twitter.com/widgets/widget_iframe.2e9f365dae390394eb8d923cba8c5b11.html?origin=https%3A%2F%2Fwww.toptal.com&amp;settingsEndpoint=https%3A%2F%2Fsyndication.twitter.com%2Fsettings" title="Twitter settings iframe" style="display: none;" frameborder="0"></iframe><div style="width:0px; height:0px; display:none; visibility:hidden;" id="batBeacon0.40719866344215305"><img style="width:0px; height:0px; display:none; visibility:hidden;" id="batBeacon0.1298405707764274" alt="" src="https://bat.bing.com/action/0?ti=5090411&amp;Ver=2&amp;mid=77d2f0d7-d98f-3c24-6ec5-f76fd021d7b9&amp;pi=0&amp;lg=de&amp;sw=1280&amp;sh=720&amp;sc=24&amp;tl=Explore%20the%20Human%20Genome%20with%20the%20SciPy%20Stack%20%7C%20Toptal&amp;p=https%3A%2F%2Fwww.toptal.com%2Fpython%2Fcomprehensive-introduction-your-genome-scipy&amp;r=&amp;lt=2220&amp;evt=pageLoad&amp;msclkid=N&amp;rn=511068" width="0" height="0"></div></body></html>