{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports (Pandas, Numpy, Spacy, NLTK, Gensim) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package wordnet to\n",
      "[nltk_data]     C:\\Users\\jacky\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import spacy\n",
    "spacy.load('en')\n",
    "from spacy.lang.en import English\n",
    "\n",
    "import nltk\n",
    "nltk.download('wordnet')\n",
    "from nltk.corpus import stopwords\n",
    "en_stops = stopwords.words('english')\n",
    "\n",
    "from nltk.corpus import wordnet as wn\n",
    "from nltk.stem.wordnet import WordNetLemmatizer\n",
    "import random\n",
    "from gensim import corpora\n",
    "import pickle\n",
    "import gensim\n",
    "import pyLDAvis.gensim\n",
    "import ensembl_rest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a dataframe object from given dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>Org_name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>CurrentID</th>\n",
       "      <th>Status</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Aliases</th>\n",
       "      <th>description</th>\n",
       "      <th>other_designations</th>\n",
       "      <th>map_location</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_nucleotide_accession.version</th>\n",
       "      <th>start_position_on_the_genomic_accession</th>\n",
       "      <th>end_position_on_the_genomic_accession</th>\n",
       "      <th>orientation</th>\n",
       "      <th>exon_count</th>\n",
       "      <th>OMIM</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>10320</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>IKZF1</td>\n",
       "      <td>CVID13, Hs.54452, IK1, IKAROS, LYF1, LyF-1, PP...</td>\n",
       "      <td>IKAROS family zinc finger 1</td>\n",
       "      <td>DNA-binding protein Ikaros|CLL-associated anti...</td>\n",
       "      <td>7p12.2</td>\n",
       "      <td>7</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>50303453.0</td>\n",
       "      <td>50405101.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>15.0</td>\n",
       "      <td>603023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>84159</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>DESRT, MRF-2, MRF2</td>\n",
       "      <td>AT-rich interaction domain 5B</td>\n",
       "      <td>AT-rich interactive domain-containing protein ...</td>\n",
       "      <td>10q21.2</td>\n",
       "      <td>10</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>61901254.0</td>\n",
       "      <td>62096948.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>12.0</td>\n",
       "      <td>608538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1053</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>CEBPE</td>\n",
       "      <td>C/EBP-epsilon, CRP1</td>\n",
       "      <td>CCAAT enhancer binding protein epsilon</td>\n",
       "      <td>CCAAT/enhancer-binding protein epsilon|CCAAT/e...</td>\n",
       "      <td>14q11.2</td>\n",
       "      <td>14</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>23117306.0</td>\n",
       "      <td>23119611.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>2.0</td>\n",
       "      <td>600749</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>1644</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>DDC</td>\n",
       "      <td>AADC</td>\n",
       "      <td>dopa decarboxylase</td>\n",
       "      <td>aromatic-L-amino-acid decarboxylase|dopa decar...</td>\n",
       "      <td>7p12.2-p12.1</td>\n",
       "      <td>7</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>50458436.0</td>\n",
       "      <td>50565460.0</td>\n",
       "      <td>minus</td>\n",
       "      <td>18.0</td>\n",
       "      <td>107930</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>7102</td>\n",
       "      <td>0</td>\n",
       "      <td>live</td>\n",
       "      <td>TSPAN7</td>\n",
       "      <td>A15, CCG-B7, CD231, DXS1692E, MRX58, MXS1, TAL...</td>\n",
       "      <td>tetraspanin 7</td>\n",
       "      <td>tetraspanin-7|CD231 antigen|T-cell acute lymph...</td>\n",
       "      <td>Xp11.4</td>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>38561478.0</td>\n",
       "      <td>38688918.0</td>\n",
       "      <td>plus</td>\n",
       "      <td>8.0</td>\n",
       "      <td>300096</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id      Org_name  GeneID  CurrentID Status  Symbol  \\\n",
       "0    9606  Homo sapiens   10320          0   live   IKZF1   \n",
       "1    9606  Homo sapiens   84159          0   live  ARID5B   \n",
       "2    9606  Homo sapiens    1053          0   live   CEBPE   \n",
       "3    9606  Homo sapiens    1644          0   live     DDC   \n",
       "4    9606  Homo sapiens    7102          0   live  TSPAN7   \n",
       "\n",
       "                                             Aliases  \\\n",
       "0  CVID13, Hs.54452, IK1, IKAROS, LYF1, LyF-1, PP...   \n",
       "1                                 DESRT, MRF-2, MRF2   \n",
       "2                                C/EBP-epsilon, CRP1   \n",
       "3                                               AADC   \n",
       "4  A15, CCG-B7, CD231, DXS1692E, MRX58, MXS1, TAL...   \n",
       "\n",
       "                              description  \\\n",
       "0             IKAROS family zinc finger 1   \n",
       "1           AT-rich interaction domain 5B   \n",
       "2  CCAAT enhancer binding protein epsilon   \n",
       "3                      dopa decarboxylase   \n",
       "4                           tetraspanin 7   \n",
       "\n",
       "                                  other_designations  map_location chromosome  \\\n",
       "0  DNA-binding protein Ikaros|CLL-associated anti...        7p12.2          7   \n",
       "1  AT-rich interactive domain-containing protein ...       10q21.2         10   \n",
       "2  CCAAT/enhancer-binding protein epsilon|CCAAT/e...       14q11.2         14   \n",
       "3  aromatic-L-amino-acid decarboxylase|dopa decar...  7p12.2-p12.1          7   \n",
       "4  tetraspanin-7|CD231 antigen|T-cell acute lymph...        Xp11.4          X   \n",
       "\n",
       "  genomic_nucleotide_accession.version  \\\n",
       "0                         NC_000007.14   \n",
       "1                         NC_000010.11   \n",
       "2                          NC_000014.9   \n",
       "3                         NC_000007.14   \n",
       "4                         NC_000023.11   \n",
       "\n",
       "   start_position_on_the_genomic_accession  \\\n",
       "0                               50303453.0   \n",
       "1                               61901254.0   \n",
       "2                               23117306.0   \n",
       "3                               50458436.0   \n",
       "4                               38561478.0   \n",
       "\n",
       "   end_position_on_the_genomic_accession orientation  exon_count    OMIM  \\\n",
       "0                             50405101.0        plus        15.0  603023   \n",
       "1                             62096948.0        plus        12.0  608538   \n",
       "2                             23119611.0       minus         2.0  600749   \n",
       "3                             50565460.0       minus        18.0  107930   \n",
       "4                             38688918.0        plus         8.0  300096   \n",
       "\n",
       "   Unnamed: 17  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df = pd.read_csv('all_genes.txt', sep='\\t', comment='#', low_memory=False, header=0)\n",
    "#df = pd.read_csv('all_genes.txt')\n",
    "df1 = pd.read_csv('all_genes.txt', error_bad_lines=False, sep='\\t', comment='#', low_memory=False, header=0)\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter Dataframe for relevant data for LDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "      <th>other_designations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IKZF1</td>\n",
       "      <td>DNA-binding protein Ikaros|CLL-associated anti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARID5B</td>\n",
       "      <td>AT-rich interactive domain-containing protein ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEBPE</td>\n",
       "      <td>CCAAT/enhancer-binding protein epsilon|CCAAT/e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDC</td>\n",
       "      <td>aromatic-L-amino-acid decarboxylase|dopa decar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSPAN7</td>\n",
       "      <td>tetraspanin-7|CD231 antigen|T-cell acute lymph...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Symbol                                 other_designations\n",
       "0   IKZF1  DNA-binding protein Ikaros|CLL-associated anti...\n",
       "1  ARID5B  AT-rich interactive domain-containing protein ...\n",
       "2   CEBPE  CCAAT/enhancer-binding protein epsilon|CCAAT/e...\n",
       "3     DDC  aromatic-L-amino-acid decarboxylase|dopa decar...\n",
       "4  TSPAN7  tetraspanin-7|CD231 antigen|T-cell acute lymph..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv('all_genes.txt',\n",
    "                         sep='\\t', comment='#', low_memory=False,\n",
    "                 usecols = [\"other_designations\",\"Symbol\"],\n",
    "                         header = 0 )\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LDA (find 5 topics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           Symbol                                 other_designations\n",
      "0           IKZF1  DNA-binding protein Ikaros|CLL-associated anti...\n",
      "1          ARID5B  AT-rich interactive domain-containing protein ...\n",
      "2           CEBPE  CCAAT/enhancer-binding protein epsilon|CCAAT/e...\n",
      "3             DDC  aromatic-L-amino-acid decarboxylase|dopa decar...\n",
      "4          TSPAN7  tetraspanin-7|CD231 antigen|T-cell acute lymph...\n",
      "5            LALL                                                NaN\n",
      "6           NALT1  MIR4674 host gene (non-protein coding)|MIR4674...\n",
      "7           BLACE                                                NaN\n",
      "8          CDKN2A  cyclin-dependent kinase inhibitor 2A|CDK4 inhi...\n",
      "9           PDE4B  cAMP-specific 3',5'-cyclic phosphodiesterase 4...\n",
      "10           IL15                                     interleukin-15\n",
      "11           TP63  tumor protein 63|amplified in squamous cell ca...\n",
      "12          GATA3  trans-acting T-cell-specific transcription fac...\n",
      "13          NCOA3  nuclear receptor coactivator 3|CBP-interacting...\n",
      "14           RYR2  ryanodine receptor 2|cardiac muscle ryanodine ...\n",
      "15          KCNA3  potassium voltage-gated channel subfamily A me...\n",
      "16        PIP4K2A  phosphatidylinositol 5-phosphate 4-kinase type...\n",
      "17          MAML2  mastermind-like protein 2|mam-2|mastermind-like 2\n",
      "18          PTPRJ  receptor-type tyrosine-protein phosphatase eta...\n",
      "19          PARD3  partitioning defective 3 homolog|CTCL tumor an...\n",
      "20          RXFP1  relaxin receptor 1|leucine-rich repeat-contain...\n",
      "21         TBXAS1  thromboxane-A synthase|TXA synthase|cytochrome...\n",
      "22          GRB10  growth factor receptor-bound protein 10|GRB10 ...\n",
      "23          ELMO1  engulfment and cell motility protein 1|ced-12 ...\n",
      "24           TP53  cellular tumor antigen p53|antigen NY-CO-13|mu...\n",
      "25          MAGI2  membrane-associated guanylate kinase, WW and P...\n",
      "26          GSTT1  glutathione S-transferase theta-1|GST class-th...\n",
      "27           RIT2  GTP-binding protein Rit2|GTP-binding protein R...\n",
      "28          RSPO2  R-spondin-2|R-spondin 2 homolog|roof plate-spe...\n",
      "29           PYGL  glycogen phosphorylase, liver form|phosphoryla...\n",
      "..            ...                                                ...\n",
      "562         UHMK1  serine/threonine-protein kinase Kist|KIS prote...\n",
      "563          HES5  transcription factor HES-5|class B basic helix...\n",
      "564        ATF7IP  activating transcription factor 7-interacting ...\n",
      "565        MIR502                                                NaN\n",
      "566        MIR448                                                NaN\n",
      "567        RASSF4  ras association domain-containing protein 4|Ra...\n",
      "568          BLID  BH3-like motif-containing cell death inducer|b...\n",
      "569       BCL2L14  apoptosis facilitator Bcl-2-like protein 14|BC...\n",
      "570         FCRL2  Fc receptor-like protein 2|IFGP family protein...\n",
      "571        ADGRF1  adhesion G-protein coupled receptor F1|G prote...\n",
      "572        UBE2Q2  ubiquitin-conjugating enzyme E2 Q2|E2 ubiquiti...\n",
      "573         FCRL1  Fc receptor-like protein 1|IFGP family protein...\n",
      "574         FHOD3  FH1/FH2 domain-containing protein 3|formactin-...\n",
      "575         TTC12  tetratricopeptide repeat protein 12|TPR repeat...\n",
      "576          TAL2  T-cell acute lymphocytic leukemia protein 2|T-...\n",
      "577        MIR612                                                NaN\n",
      "578        CREBL2  cAMP-responsive element-binding protein-like 2...\n",
      "579        CLECL1  C-type lectin-like domain family 1|DC-associat...\n",
      "580        SLX4IP                                     protein SLX4IP\n",
      "581         HMHB1  minor histocompatibility protein HB-1|minor hi...\n",
      "582        ZNF296    zinc finger protein 296|zinc finger protein 342\n",
      "583       MIR520H                                       hsa-mir-520h\n",
      "584         RCSD1  capZ-interacting protein|RCSD domain-containin...\n",
      "585  LOC107980440  ABL proto-oncogene 1, non-receptor tyrosine ki...\n",
      "586  LOC107963955  BCR p210 Philadelphia chromosome recombination...\n",
      "587  LOC107963951  BCR p190 Philadelphia chromosome recombination...\n",
      "588      NEWENTRY                                                NaN\n",
      "589  LOC107648866                                                NaN\n",
      "590         HOXA@                                homeo box A cluster\n",
      "591           E2A                single-stranded DNA-binding protein\n",
      "\n",
      "[592 rows x 2 columns]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "object of type 'int' has no len()",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-6d0917fb32b5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     41\u001b[0m \u001b[0mtext_data\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     42\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 43\u001b[1;33m     \u001b[0mtokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mprepare_text_for_lda\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     44\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mrandom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m.99\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     45\u001b[0m         \u001b[0mtext_data\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtokens\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-5-6d0917fb32b5>\u001b[0m in \u001b[0;36mprepare_text_for_lda\u001b[1;34m(text)\u001b[0m\n\u001b[0;32m     32\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     33\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mprepare_text_for_lda\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 34\u001b[1;33m     \u001b[0mtokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtokenize\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     35\u001b[0m     \u001b[0mtokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mtoken\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mtoken\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mtokens\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtoken\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m4\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     36\u001b[0m     \u001b[0mtokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mtoken\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mtoken\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mtokens\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mtoken\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mcustom_stop\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-5-6d0917fb32b5>\u001b[0m in \u001b[0;36mtokenize\u001b[1;34m(text)\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mtokenize\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mlda_tokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 6\u001b[1;33m     \u001b[0mtokens\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      7\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mtoken\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mtokens\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mtoken\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0morth_\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misspace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\spacy\\language.py\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, text, disable)\u001b[0m\n\u001b[0;32m    341\u001b[0m             \u001b[1;33m(\u001b[0m\u001b[1;34m'An'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'NN'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    342\u001b[0m         \"\"\"\n\u001b[1;32m--> 343\u001b[1;33m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmax_length\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    344\u001b[0m             raise ValueError(Errors.E088.format(length=len(text),\n\u001b[0;32m    345\u001b[0m                                                 max_length=self.max_length))\n",
      "\u001b[1;31mTypeError\u001b[0m: object of type 'int' has no len()"
     ]
    }
   ],
   "source": [
    "parser = English()\n",
    "\n",
    "# Divide data into atomic elements\n",
    "def tokenize(text):\n",
    "    lda_tokens = []\n",
    "    tokens = parser(text)\n",
    "    for token in tokens:\n",
    "        if token.orth_.isspace():\n",
    "            continue\n",
    "        elif token.like_url:\n",
    "            lda_tokens.append('URL')\n",
    "        elif token.orth_.startswith('@'):\n",
    "            lda_tokens.append('SCREEN_NAME')\n",
    "        else:\n",
    "            lda_tokens.append(token.lower_)\n",
    "    return lda_tokens\n",
    "\n",
    "# Lemmatize words \n",
    "def get_lemma(word):\n",
    "    lemma = wn.morphy(word)\n",
    "    if lemma is None:\n",
    "        return word\n",
    "    else:\n",
    "        return lemma\n",
    "    \n",
    "def get_lemma2(word):\n",
    "    return WordNetLemmatizer().lemmatize(word)\n",
    "\n",
    "custom_stop = ['sapiens', 'protein', 'homo', '9606', 'single', 'minus', 'plus', 'homeobox', 'human', 'binding'\n",
    "              'strand', 'nc_001460.1']\n",
    "custom_stop.extend(en_stops)\n",
    "\n",
    "def prepare_text_for_lda(text):\n",
    "    tokens = tokenize(text)\n",
    "    tokens = [token for token in tokens if len(token) > 4]\n",
    "    tokens = [token for token in tokens if token not in custom_stop]\n",
    "    tokens = [get_lemma(token) for token in tokens]\n",
    "    return tokens\n",
    "\n",
    "print(df2)\n",
    "text_data = []\n",
    "for i in range(0, len(df2)-1):\n",
    "    tokens = prepare_text_for_lda(i)\n",
    "    if random.random() > .99:\n",
    "        text_data.append(tokens)\n",
    "            \n",
    "# create a dictionary from the data\n",
    "# convert it into a 'bag of words' corpus\n",
    "dictionary = corpora.Dictionary(text_data)\n",
    "corpus = [dictionary.doc2bow(text) for text in text_data]\n",
    "pickle.dump(corpus, open('corpus.pkl', 'wb'))\n",
    "dictionary.save('dictionary.gensim')\n",
    "\n",
    "# define number of topics\n",
    "NUM_TOPICS = 5\n",
    "ldamodel = gensim.models.ldamodel.LdaModel(corpus, num_topics = NUM_TOPICS, id2word=dictionary, \n",
    "                                           passes=15)\n",
    "ldamodel.save('model5.gensim')\n",
    "\n",
    "topics = ldamodel.print_topics(num_words=4)\n",
    "for topic in topics:\n",
    "    print(topic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LDA with 3 topics "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ldamodel = gensim.models.ldamodel.LdaModel(corpus, num_topics = 3, id2word=dictionary, passes=50, update_every = 3)\n",
    "ldamodel.save('model3.gensim')\n",
    "topics = ldamodel.print_topics(num_words=6)\n",
    "for topic in topics:\n",
    "    print(topic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LDA with 10 topics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ldamodel = gensim.models.ldamodel.LdaModel(corpus, num_topics = 10, id2word=dictionary, passes=10, update_every = 3)\n",
    "ldamodel.save('model10.gensim')\n",
    "topics = ldamodel.print_topics(num_words=6)\n",
    "for topic in topics:\n",
    "    print(topic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Draw the distribution using pyLDAvis  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dictionary = gensim.corpora.Dictionary.load('dictionary.gensim')\n",
    "corpus = pickle.load(open('corpus.pkl', 'rb'))\n",
    "lda = gensim.models.ldamodel.LdaModel.load('model5.gensim')\n",
    "\n",
    "lda_display = pyLDAvis.gensim.prepare(lda, corpus, dictionary, sort_topics=False)\n",
    "pyLDAvis.display(lda_display)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lda3 = gensim.models.ldamodel.LdaModel.load('model3.gensim')\n",
    "lda_display3 = pyLDAvis.gensim.prepare(lda3, corpus, dictionary, sort_topics=False)\n",
    "pyLDAvis.display(lda_display3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lda10 = gensim.models.ldamodel.LdaModel.load('model10.gensim')\n",
    "lda_display10 = pyLDAvis.gensim.prepare(lda10, corpus, dictionary, sort_topics=False)\n",
    "pyLDAvis.display(lda_display10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
